---
title: "Run the isotwas pipeline"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Run the isotwas pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

## Introduction

This module will cover how to use functions from `isotwas` to train an multivariate `isotwas` model for isoform-level expression. Then, we will run summary statistics-based disease mapping using the stage-wise hypothesis testing procedure. Lastly, we will run probabilistic fine-mapping. Along the way, we'll be covering some practical considerations for computation and feature selection.

```{r setup, echo=F, message=F}
library(isotwas)
```

## Train predictive model

To train a multivariate model, we need:

1.  a matrix of dosages of SNPs, and

2.  a matrix of isoform-level expressions for a given gene.

Isoform expression estimated from pseudo-alignment methods like [`salmon`](https://combine-lab.github.io/salmon/) and [`kallisto`](https://pachterlab.github.io/kallisto/) provide bootstrapped estimates expression across a number of iterations. This information can be included into model training, as illustrated below.

[**Practical consideration**]{.ul}: Ideally, to reduce your memory usage, we recommend storing your genetic data as `PLINK` format data (ideally as `.bed`/`bim`/`fam` or `.pgen`/`.pvar`/`.psam`). For a given gene, use `PLINK` to [select only SNPs](https://www.cog-genomics.org/plink/2.0/filter) on the same chromosome and within 1 Megabase of the gene body. Then, you can use the [`bigsnpr` package](https://privefl.github.io/bigsnpr/) read in this "condensed" file set.

For this vignette, we'll use some toy genetic and transcriptomic data provided.

```{r generate_genetic_data}
train_bed = system.file("extdata", "train.bed", package = "isotwas")
snps = bigsnpr::snp_attach(bigsnpr::snp_readBed2(train_bed,
                                                 backingfile = tempfile()))
dim(snps$genotypes[])

isoform_mat = readRDS(system.file("extdata", "isoform_exp.RDS", package = "isotwas"))
dim(isoform_mat)
```

This reads in a SNP matrix with dimensions $300 \times 4542$, with samples along the rows and SNPs along the columns and a isoform-level expression matrix of dimension $300 \times 3$, with isoforms along the columns.

[**Practical consideration**]{.ul}: Though this is not a requirement for the isotwas pipeline, we recommend a feature selection step based on cis-heritability of isoform-level expression. Prior of model training for isoforms of a given gene, we recommend that users compute the heritability of all expressed isoforms of a gene using [`GCTA`](https://yanglab.westlake.edu.cn/software/gcta/#Overview) (ideally the [`gcta-nr-robust` executable in `FUSION`](http://gusevlab.org/projects/fusion/)). We recommend users only train models for isoforms that have positive heritability at nominal $P < 0.05$ or $0.01$.

[**Practical consideration**]{.ul}: Prior to heritability analysis and model training, it is imperative that your expression data is residualized by relevant covariates (i.e. principal components of genotype matrix, hidden covariates or PEER factors of the expression matrix, clinical/demographic covariates). A general rule of thumb is if you would include the covariate in an eQTL/isoform-level eQTL analysis, you should include it in the residualization process.

Let's jitter our expression matrix a little to generate 10 bootstrapped estimates of the matrix. Keep in mind that these replicates, theoretically, should not affect the effect size estimates but will help in seeding the estimation process of the correlation between isoforms/error in the multivariate model.

```{r bootstrap}
boot_list = lapply(1:10,function(i){jitter(isoform_mat)}) 
isoform_mat_rep = rlist::list.rbind(boot_list) 
rownames(isoform_mat_rep) = rep(paste0('Sample',1:nrow(snps$fam)),10) 
colnames(isoform_mat_rep) = paste0('Isoform',1:ncol(isoform_mat_rep)) 
dim(isoform_mat_rep)
```

Now, we can train the model using the `compute_isotwas()` function. Make sure the SNP matrix has clear labels for the SNP identifiers that are relevant to your study. This is a large SNP matrix which will take a few minutes to run on your local machine.

```{r train, message = F, warning=FALSE, eval=F}
snp_mat = as.matrix(snps$genotypes[])
colnames(snp_mat) = snps$map$marker.ID
rownames(snp_mat) = snps$fam$sample.ID
isotwas_model = compute_isotwas(X = snp_mat, 
                                Y = isoform_mat, 
                                Y.rep = isoform_mat_rep,
                                R = 10, 
                                id = rownames(isoform_mat_rep), 
                                omega_est = 'replicates', 
                                omega_nlambda = 5, 
                                method = c('mrce_lasso', 
                                           'multi_enet', 
                                           'univariate',
                                           'mvsusie'),
                                predict_nlambda = 10, 
                                family = 'gaussian', 
                                scale = FALSE, 
                                alpha = 0.5, 
                                nfolds = 5, 
                                verbose = TRUE, 
                                tx_names = paste0('Isoform',1:3), 
                                seed = 1789, 
                                run_all = FALSE, 
                                return_all = TRUE)
```

A few notes on the options in the function:

-   By selecting `omega_est = 'replicates'` , we are used the bootstrapped replicates to estimate the correlation between isoforms to seed any methods that require this matrix. The alternative is `omega_est = 'mean'` , which will use whatever matrix you submit for `Y`.

-   `omega_nlambda` selects the number of steps between fully dense to fully sparse in the correlation matrix estimation. We recommend 5-10 as the default here.

-   There are 6 total methods to predict expression. However, through simulations and real data examples, we recommend the four shown here as the default in the `method` option.

-   `predict_nlambda` is the number of penalty variables that are considered from the `mrce_lasso` method. We recommend 10-20 for this variable.

-   `family` and `alpha` are for `multi_enet` and `univariate` to determine how elastic net regression is run. The default is that the response follows a Normal distribution and the mixing parameter is set to 0.5.

-   If `run_all = TRUE`, then all 6 methods will be run.

-   If `return_all = TRUE`, then predicted values for each isoform and predictive performance for all methods and all isoforms will be outputted.

We can convert the model from a list (the output from `compute_isotwas()`) to a tibble or data.frame using `convert_model(isotwas_model)`.

Let's take a look at what a sample model looks like:

```{r sample}
model_isotwas = readRDS(system.file("extdata", "model_example.RDS", package = "isotwas"))
class(model_isotwas)
length(model_isotwas)
names(model_isotwas)
model_isotwas$Model[[1]]
```

This model object is a list with a `Model` and a `R2` slot. The `Model` slot stores a list of models for each isoform.

We can convert this model from a list to a `tibble` using the `convert_model()` function. Make sure you have an annotation object to have position and REF/ALT allele information.

```{r convert}
model_tsv = convert_model(model_isotwas,
                          snp_annot = snps$map,
                          snp_var = 'marker.ID')
model_tsv
```

## Run stage-wise trait mapping

Let's load up the sample GWAS data that's provided and read in the reference linkage disequilibrium data. Again, these GWAS summary statistic files and LD reference files are large and unwieldy. Try to use bash or `PLINK` commands to cut down the file to the variants you need. You need to have standardized effect sizes for the GWAS (i.e. `beta`/`SE`). You only need to provide a numeric matrix for the LD reference in the burden test function. Here, we use `bigsnpr` to calculate the reference LD.

```{r gwas}
gwas = read.table(system.file("extdata", "gwas.txt", package = "isotwas"),header = T)
head(gwas)

ld_bed = system.file("extdata", "ld.bed", package = "isotwas")
ld_ref = bigsnpr::snp_attach(bigsnpr::snp_readBed2(ld_bed,
                                                   backingfile = tempfile()))
ld_cor = bigsnpr::snp_cor(ld_ref$genotypes)
colnames(ld_cor) = rownames(ld_cor) = ld_ref$map$marker.ID
```

Let's now generate TWAS Z-scores for each isoform using `burdenTest()` modified from the [`MOSTWAS` package](https://github.com/bhattacharya-a-bt/MOSTWAS). Ensure that you have a column named `Chromosome`, `Position`, `A1` (alternative allele) and `A2` (reference allele) in your model tibble.

```{r runZ}
model_tsv$Chromosome = model_tsv$chromosome
model_tsv$Position = model_tsv$physical.pos
model_tsv$A1 = model_tsv$allele1
model_tsv$A2 = model_tsv$allele2
res.df = data.frame(Gene = c(),
                    Transcript = c(),
                    Z = c(),
                    P = c(),
                    permute.P = c(),
                    topSNP = c(),
                    topSNP.P = c())
for (isoform in unique(model_tsv$Transcript)){
  
  model_this = subset(model_tsv,
                      Transcript == isoform)
  gwas_cur = subset(gwas,SNP %in% model_this$SNP)
  gwas_cur = gwas_cur[match(model_this$SNP,
                            gwas_cur$SNP),]
  res.df = rbind(res.df,
                 burdenTest(mod = model_this,
                            ld = ld_cor,
                            gene = 'GeneExample',
                            sumStats = gwas_cur,
                            chr = 'Chromosome',
                            pos = 'Position',
                            a1 = 'A1',
                            a2 = 'A2',
                            Z = 'Z',
                            beta = 'Beta',
                            se = 'SE',
                            R2cutoff = 0.01,
                            alpha = 0.25,
                            nperms = 1e2,
                            usePos = F))
  
}
res.df
```

Now, we have individual isoform-level associations for all isoforms of our example gene. We can run a screening-level test to test whether at least one of the isoforms of the gene is significantly associated with the trait using `p_screen(res.df$P,mode = 'P')`. For a panel of genes, we run false discovery control on these screening P-values. If a gene has a small enough P-value on the screening level, we run `p_confirm(res.df$P)` to control the family-wise error rate within each gene family. This will provide a set of isoforms that pass stage-wise hypothesis testing.

We recommend permutation testing on the isoform-level to make sure you aren't proritizing an isoform-trait association only because of strong GWAS or strong QTL effects; read [Gusev et al 2016](https://www.nature.com/articles/ng.3506) for more information. `burdenTest()` will automatically run permutations if the nominal TWAS P-value for an isoform is under the specified `alpha`.

## Probabilistic fine-mapping

We have modified functions from Mancuso et al 2019's [FOCUS software](https://github.com/bogdanlab/focus); here's a [link to the paper](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6619422/). All you need for this is the models for the transcript you're interested in, an LD reference matrix, and the TWAS Z-scores.

We'll generate one matrix: `Omega` with 3 columns for each isoform and SNPs along the rows. And then we will make sure the LD matrix aligns with the SNPs in `Omega`.

```{r focus}
Omega = matrix(nrow = length(unique(model_tsv$SNP)),
               ncol = nrow(res.df))
colnames(Omega) = res.df$Transcript
rownames(Omega) = unique(model_tsv$SNP)
V = ld_cor[rownames(Omega),rownames(Omega)]
for (i in 1:ncol(Omega)){
  for (j in 1:nrow(Omega)){
    
    mod_this = subset(model_tsv,Transcript == colnames(Omega)[i])
    Omega[j,i] = ifelse(rownames(Omega)[j] %in% mod_this$SNP,
                        mod_this$Weight[mod_this$SNP == rownames(Omega)[j]],
                        0)
    
    
  }
}

res.df = twas_finemap(res.df = res.df,
                      z = 'Z',
                      Omega = Omega,
                      V = V, 
                      max_enum = 3,
                      cutoff = .9)
res.df
```
