---
title: "Train an isotwas model"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Train an isotwas model}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

This module will cover how to use functions from `isotwas` to train an multivariate isotwas model for isoform-level expression. Along the way, we'll be covering some practical considerations for computation and feature selection.

```{r setup, echo=F, message=F}
library(isotwas)
```

To train a multivariate model, we need:

1.  a matrix of dosages of SNPs, and

2.  a matrix of isoform-level expressions for a given gene.

Isoform expression estimated from pseudo-alignment methods like [`salmon`](https://combine-lab.github.io/salmon/) and [`kallisto`](https://pachterlab.github.io/kallisto/) provide bootstrapped estimates expression across a number of iterations. This information can be included into model training, as illustrated below.

[**Practical consideration**]{.ul}: Ideally, to reduce your memory usage, we recommend storing your genetic data as `PLINK` format data (ideally as `.bed`/`bim`/`fam` or `.pgen`/`.pvar`/`.psam`). For a given gene, use `PLINK` to [select only SNPs](https://www.cog-genomics.org/plink/2.0/filter) on the same chromosome and within 1 Megabase of the gene body. Then, you can use the [`bigsnpr` package](https://privefl.github.io/bigsnpr/) read in this "condensed" file set.

For this vignette, we'll generate some toy genetic and transcriptomic data with $p = 3$ isoforms for a gene and $k = 500$ SNPs within 1 Megabase of the gene across $n = 200$ samples.
