<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>Run isoTWAS associations</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Run isoTWAS associations</h1>



<p>This module will cover how to use functions from <code>isotwas</code>
to map isoform-level trait associations using a two-step testing
framework. Then, we will run probabilistic fine-mapping to account of
linkage disequilibrium (LD) in a locus with overlapping isoforms that
are associated with a trait. Along the way, we’ll be covering some
practical considerations for computation and feature selection.</p>
<p>For this, you will need:</p>
<ol style="list-style-type: decimal">
<li><p>a set of isoTWAS models,</p></li>
<li><p>an appropriate LD reference panel, and</p></li>
<li><p>GWAS summary statistics for your trait of interest.</p></li>
</ol>
<p>Let’s talk a little about each of these three components.</p>
<div id="pre-computed-models-and-ld-matrix" class="section level2">
<h2>Pre-computed models and LD matrix</h2>
<p>You can always use your own set of models that you have trained on
your own isoform expression QTL data (paired genotype and RNA-seq data).
For convenience, we have trained and made publicly available isoTWAS
models for 48 tissues from the <a href="https://gtexportal.org/home/">Genotype-Tissue Expression Project
(GTEx)</a>, adult brain cortex from <a href="https://www.nimhgenetics.org/resources/psychencode">PsychENCODE</a>
and <a href="https://www.nia.nih.gov/research/amp-ad">AMP-AD</a>, and
developmental brain cortex from PsychENCODE. There are Zenodo
repositories for the <a href="https://zenodo.org/record/8047940">GTEx</a>, <a href="https://zenodo.org/record/8048198">adult cortex</a>, and <a href="https://zenodo.org/record/8048137">developmental cortex</a>
models. Please cite our manuscript and the associated Zenodo DOI if you
use these models.</p>
<p>To download an entire Zenodo repository, you may use the Python
package <code>Zenodo_get</code> (<a href="https://zenodo.org/record/1261813" class="uri">https://zenodo.org/record/1261813</a>). To install and use,
here are some sample commands to directly download all of the adult
cortex repository off Zenodo:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install zenodo-get</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="ex">zenodo_get</span> 8048198</span></code></pre></div>
<p>In this tutorial, we are only using isoTWAS models for 3 genes:
<em>CDC42</em> (<em>ENSG00000070831</em>), <em>WNT4</em>
(<em>ENSG00000162552</em>), and <em>CDC42-IT1</em>
(<em>ENSG00000230068</em>). Let’s take a look at what the models look
like, which are saved as <code>.RDS</code> files.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>model_test <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">&quot;ENSG00000070831_isoTWAS.RDS&quot;</span>,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(model_test)</span></code></pre></div>
<pre><code>##           Feature        SNP Chromosome Position Build ALT REF        Weight
## 2 ENST00000344548 rs10917023          1 21599110  hg38   A   G  0.0257706583
## 3 ENST00000344548 rs10917145          1 22086152  hg38   A   G -0.2503237794
## 4 ENST00000344548 rs11584687          1 22550747  hg38   A   G  0.0200388736
## 5 ENST00000344548 rs12082914          1 21493245  hg38   G   T  0.0167839153
## 6 ENST00000344548 rs12118362          1 21445504  hg38   G   A  0.0133152139
## 7 ENST00000344548 rs12127343          1 21543280  hg38   A   G  0.0007759457
##           R2            Gene
## 2 0.09314035 ENSG00000070831
## 3 0.09314035 ENSG00000070831
## 4 0.09314035 ENSG00000070831
## 5 0.09314035 ENSG00000070831
## 6 0.09314035 ENSG00000070831
## 7 0.09314035 ENSG00000070831</code></pre>
<p>If you train your own models, ensure that your model files contain
columns for the:</p>
<ol style="list-style-type: decimal">
<li><p>isoform name,</p></li>
<li><p>SNP name or rsID,</p></li>
<li><p>chromosome of the SNP,</p></li>
<li><p>position of the SNP,</p></li>
<li><p>alternative allele,</p></li>
<li><p>reference allele, and</p></li>
<li><p>the predictive weight for the isoform.</p></li>
</ol>
<p>For reference, our models include the genome build on which the model
was trained, so users can align their GWAS to the models based on genome
build.</p>
<p>We also include in-sample LD matrices for each isoTWAS model. These
are simply <code>.RDS</code> files that contain a square matrix with the
LD between the SNPs in the correspondin row and column.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>LD_test <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>                                 <span class="st">&quot;ENSG00000070831_LDMatrix.RDS&quot;</span>,</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>                                 <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">as.matrix</span>(LD_test)[,<span class="dv">1</span><span class="sc">:</span><span class="dv">10</span>])</span></code></pre></div>
<pre><code>##            rs10916927  rs6658526  rs1354792 rs12567861 rs10916930  rs7521711
## rs10916927  1.0000000  0.9440432  0.8544395 -0.6393393 -0.6405133  0.8536513
## rs6658526   0.9440432  1.0000000  0.8005756 -0.5896286 -0.5924963  0.7998371
## rs1354792   0.8544395  0.8005756  1.0000000 -0.7649469 -0.7657869  0.9990775
## rs12567861 -0.6393393 -0.5896286 -0.7649469  1.0000000  0.9971883 -0.7642413
## rs10916930 -0.6405133 -0.5924963 -0.7657869  0.9971883  1.0000000 -0.7650804
## rs7521711   0.8536513  0.7998371  0.9990775 -0.7642413 -0.7650804  1.0000000
##             rs4654899   rs960564  rs6663452  rs7532667
## rs10916927  0.9097877  0.9966544  0.9501309  0.9971286
## rs6658526   0.8503948  0.9427488  0.8973393  0.9458899
## rs1354792   0.9361329  0.8530981  0.8951509  0.8538566
## rs12567861 -0.7173327 -0.6390579 -0.6724022 -0.6393520
## rs10916930 -0.7179086 -0.6402319 -0.6737519 -0.6405219
## rs7521711   0.9371127  0.8532494  0.8952461  0.8530689</code></pre>
<p>Note that the column and row names for the matrix correspond to the
rsIDs in the model.</p>
<p><u><strong>Practical consideration</strong></u>: In many settings,
you will not have a corresponding file that contains the LD matrix for
the SNPs at the locus. In this case, we suggest either one of two
approaches. First, you can generate the LD matrix from the data where
the model was trained, if you have access to the individual-level
genotypes for this data. Alternatively, you can use an LD reference
panel that is well-aligned to the genetic ancestry of the training
dataset. This LD reference panel is usually <a href="https://www.internationalgenome.org/">1000 Genomes</a>, since it’s
publicly available. You can use <code>bigsnpr::snp_cor</code> or
<code>bigsnpr::bed_cor</code> to generate similar-looking LD matrices,
given you have the PLINK style files
(<code>.bed</code>/<code>.bim</code>/<code>.fam</code>) for the LD
reference.</p>
</div>
<div id="preparing-the-gwas-summary-statistics" class="section level2">
<h2>Preparing the GWAS summary statistics</h2>
<p>The last file you will need is the GWAS summary statistics for your
trait of interest. In this tutorial, we are using the GWAS summary
statistics for schizophrenia, made available by the Psychiatric Genomics
Consortium. We are specifically using the GWAS summary statistics from
European-ancestry individual. You can find these files here: <a href="https://figshare.com/articles/dataset/scz2022/19426775" class="uri">https://figshare.com/articles/dataset/scz2022/19426775</a>.</p>
<p>Since GWAS summary statistics often have varied column names, we use
a munging script to clean the data. In addition, we subset the GWAS
summary statistics to only SNPs annotated in <a href="https://www.nature.com/articles/nature02168">HapMap3</a> to ensure
that there is maximal overlap. This munging script is available <a href="https://github.com/bhattacharya-a-bt/isotwas_manu_scripts/blob/98a76279037003e0e30de964b26fe7d5a49961df/Real%20Data/Association%20Testing/MungeSumStats/Snakefile">here</a>,
courtesy of Michael Margolis.</p>
<p>In this tutorial, we have subsetted only to the SNPs that are within
the <em>SCN2A</em>/<em>SCN3A</em> locus. Let’s take a look at this
munged file.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a>gwas <span class="ot">=</span> <span class="fu">read.table</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>                              <span class="st">&quot;tutorial_SCZ_2022.hm3_filter.sumstats.gz&quot;</span>,</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>                              <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>),</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>                  <span class="at">header=</span>T)</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gwas)</span></code></pre></div>
<pre><code>##          SNP A1 A2      Z        N
## 1 rs10916927  T  C  3.108 146347.8
## 2  rs6658526  C  T  2.954 146347.8
## 3  rs1354792  T  C  1.636 146347.8
## 4 rs12567861  C  T -1.641 146347.8
## 5 rs10916930  T  C -1.684 146347.8
## 6  rs7521711  A  G  1.627 146347.8</code></pre>
<p>This munged summary statistics table has only 5 columns:</p>
<ol style="list-style-type: decimal">
<li><p>The SNP rsID,</p></li>
<li><p>the alternative allele,</p></li>
<li><p>the reference allele,</p></li>
<li><p>the Z-score (beta value divided by standard error) for the GWAS
association, and</p></li>
<li><p>the effective sample size.</p></li>
</ol>
<p>We advise you use the munging script to format your summary
statistics files. At the very least, please ensure that your summary
statistics file contains a column for the SNP identifier, the
alternative allele, and either the Z-score or the beta value and
standard error for each SNP.</p>
</div>
<div id="generating-nominal-z-scores" class="section level2">
<h2>Generating nominal Z-scores</h2>
<p>The first step for isoTWAS trait-mapping is to generate the
isoform-level Z-scores using the weighted burden test. In general,
you’ll have a vector of gene names with corresponding isoTWAS model
files and/or LD files. The following code snippet loops through this
vector of genes, reads in the model file and LD file, and then computes
a Z-score for each isoform of the gene that was well-predicted in
cross-validation.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a>gene_names <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;ENSG00000070831&#39;</span>,</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>               <span class="st">&#39;ENSG00000162552&#39;</span>,</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>               <span class="st">&#39;ENSG00000230068&#39;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>out_df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">Gene =</span> <span class="fu">c</span>(),</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>                    <span class="at">Feature =</span> <span class="fu">c</span>(),</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>                    <span class="at">Z =</span> <span class="fu">c</span>(),</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>                    <span class="at">P =</span> <span class="fu">c</span>(),</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>                    <span class="at">permute.P =</span> <span class="fu">c</span>(),</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a>                    <span class="at">topSNP =</span> <span class="fu">c</span>(),</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>                    <span class="at">topSNP.P =</span> <span class="fu">c</span>())</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (gene <span class="cf">in</span> gene_names){</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  model <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">paste0</span>(gene,<span class="st">&quot;_isoTWAS.RDS&quot;</span>),</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>                              <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a>  ld <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>                           <span class="fu">paste0</span>(gene,<span class="st">&quot;_LDMatrix.RDS&quot;</span>),</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>                           <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (tx <span class="cf">in</span> <span class="fu">unique</span>(model<span class="sc">$</span>Feature)){</span>
<span id="cb8-22"><a href="#cb8-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-23"><a href="#cb8-23" aria-hidden="true" tabindex="-1"></a>    sumstats.cur <span class="ot">=</span> <span class="fu">subset</span>(gwas,SNP <span class="sc">%in%</span> <span class="fu">subset</span>(model, Feature <span class="sc">==</span> tx)<span class="sc">$</span>SNP)</span>
<span id="cb8-24"><a href="#cb8-24" aria-hidden="true" tabindex="-1"></a>    tx_df <span class="ot">=</span> isotwas<span class="sc">::</span><span class="fu">burdenTest</span>(<span class="at">mod =</span> <span class="fu">subset</span>(model, Feature <span class="sc">==</span> tx),</span>
<span id="cb8-25"><a href="#cb8-25" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ld =</span> ld,</span>
<span id="cb8-26"><a href="#cb8-26" aria-hidden="true" tabindex="-1"></a>                       <span class="at">gene =</span> gene,</span>
<span id="cb8-27"><a href="#cb8-27" aria-hidden="true" tabindex="-1"></a>                       <span class="at">sumStats =</span> sumstats.cur,</span>
<span id="cb8-28"><a href="#cb8-28" aria-hidden="true" tabindex="-1"></a>                       <span class="at">chr =</span> <span class="st">&#39;Chromosome&#39;</span>,</span>
<span id="cb8-29"><a href="#cb8-29" aria-hidden="true" tabindex="-1"></a>                       <span class="at">pos =</span> <span class="st">&#39;Position&#39;</span>,</span>
<span id="cb8-30"><a href="#cb8-30" aria-hidden="true" tabindex="-1"></a>                       <span class="at">a1 =</span> <span class="st">&#39;A1&#39;</span>,</span>
<span id="cb8-31"><a href="#cb8-31" aria-hidden="true" tabindex="-1"></a>                       <span class="at">a2 =</span> <span class="st">&#39;A2&#39;</span>,</span>
<span id="cb8-32"><a href="#cb8-32" aria-hidden="true" tabindex="-1"></a>                       <span class="at">a1_mod =</span> <span class="st">&#39;ALT&#39;</span>,</span>
<span id="cb8-33"><a href="#cb8-33" aria-hidden="true" tabindex="-1"></a>                       <span class="at">a2_mod =</span> <span class="st">&#39;REF&#39;</span>,</span>
<span id="cb8-34"><a href="#cb8-34" aria-hidden="true" tabindex="-1"></a>                       <span class="at">snpName =</span> <span class="st">&#39;SNP&#39;</span>,</span>
<span id="cb8-35"><a href="#cb8-35" aria-hidden="true" tabindex="-1"></a>                       <span class="at">Z =</span> <span class="st">&#39;Z&#39;</span>,</span>
<span id="cb8-36"><a href="#cb8-36" aria-hidden="true" tabindex="-1"></a>                       <span class="at">beta =</span> <span class="cn">NULL</span>,</span>
<span id="cb8-37"><a href="#cb8-37" aria-hidden="true" tabindex="-1"></a>                       <span class="at">se =</span> <span class="cn">NULL</span>,</span>
<span id="cb8-38"><a href="#cb8-38" aria-hidden="true" tabindex="-1"></a>                       <span class="at">featureName =</span> <span class="st">&#39;Feature&#39;</span>,</span>
<span id="cb8-39"><a href="#cb8-39" aria-hidden="true" tabindex="-1"></a>                       <span class="at">R2cutoff =</span> .<span class="dv">01</span>,</span>
<span id="cb8-40"><a href="#cb8-40" aria-hidden="true" tabindex="-1"></a>                       <span class="at">alpha =</span> <span class="fl">1e-3</span>,</span>
<span id="cb8-41"><a href="#cb8-41" aria-hidden="true" tabindex="-1"></a>                       <span class="at">nperms =</span> <span class="dv">1000</span>,</span>
<span id="cb8-42"><a href="#cb8-42" aria-hidden="true" tabindex="-1"></a>                       <span class="at">usePos =</span> F)</span>
<span id="cb8-43"><a href="#cb8-43" aria-hidden="true" tabindex="-1"></a>    out_df <span class="ot">=</span> <span class="fu">rbind</span>(out_df,tx_df)</span>
<span id="cb8-44"><a href="#cb8-44" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb8-45"><a href="#cb8-45" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb8-46"><a href="#cb8-46" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb8-47"><a href="#cb8-47" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-48"><a href="#cb8-48" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(out_df)</span></code></pre></div>
<pre><code>##              Gene         Feature         Z            P  permute.P     topSNP
## 1 ENSG00000070831 ENST00000344548 -3.239210 0.0011986108 1.00000000  rs2143103
## 2 ENSG00000070831 ENST00000315554  1.268410 0.2046514827 1.00000000 rs11582542
## 3 ENSG00000070831 ENST00000656825  3.268385 0.0010816323 1.00000000 rs11582542
## 4 ENSG00000070831 ENST00000498236 -2.906460 0.0036554386 1.00000000 rs11582542
## 5 ENSG00000070831 ENST00000400259  3.344840 0.0008232993 0.01098901 rs11582542
## 6 ENSG00000070831 ENST00000667384  0.575199 0.5651567515 1.00000000 rs11582542
##       topSNP.P
## 1 0.0004331767
## 2 0.0003396706
## 3 0.0003396706
## 4 0.0003396706
## 5 0.0003396706
## 6 0.0003396706</code></pre>
<p>The <code>out_df</code> variable contains a data frame with 7
columns:</p>
<ol style="list-style-type: decimal">
<li><p>the gene name,</p></li>
<li><p>the isoform name,</p></li>
<li><p>the isoform-level Z-score for the trait association,</p></li>
<li><p>the P-value corresponding to this Z-score,</p></li>
<li><p>the permutation P-value (more on this below),</p></li>
<li><p>the top SNP based on GWAS P-value within 1 Mb of the gene
locus,</p></li>
<li><p>and the P-value corresponding to this GWAS SNP.</p></li>
</ol>
<p>The permutation test test assesses how much signal is added by
isoform expression, given the GWAS architecture of the locus, and
controls for large LD blocks. We shuffle the SNP-to-isoform weights to
generate a null distribution and use this null to generate the
permutation P-value. We only run this permutation test if the nominal
P-value from the isoform-level Z-score is less than <code>alpha</code>,
as defined in the <code>burdenTest</code> function.</p>
</div>
<div id="stage-wise-hypothesis-test" class="section level2">
<h2>Stage-wise hypothesis test</h2>
<p><code>isoTWAS</code> employs a two-step testing framework. First, we
control from false discovery rate (FDR) across gene families, using
either Benjamini-Hochberg FDR control (recommended) or Bonferroni. Next,
for genes that pass FDR control (adjusted P-values for the genes are
lower than our defined significance threshold), we drop down to the
isoform level and run family-wise error rate (FWER) control using
Shaffer’s modified sequentially rejective Bonferroni procedure. Here, we
are controlling both the FDR (<code>alpha1</code>) and FWER
(<code>alpha2</code>) to 0.05.</p>
<p>Let’s run the first stage now: gene-level FDR control.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="sc">!</span><span class="fu">require</span>(<span class="st">&quot;tidyverse&quot;</span>, <span class="at">quietly =</span> <span class="cn">TRUE</span>)){</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true" tabindex="-1"></a>    <span class="fu">install.packages</span>(<span class="st">&quot;tidyverse&quot;</span>,</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true" tabindex="-1"></a>                     <span class="at">repos =</span> <span class="st">&#39;https://archive.linux.duke.edu/cran/&#39;</span>,</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true" tabindex="-1"></a>                     <span class="at">dependencies =</span> T)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true" tabindex="-1"></a>  }</span></code></pre></div>
<pre><code>## ── Attaching packages ─────────────────────────────────────── tidyverse 1.3.2 ──
## ✔ ggplot2 3.4.2     ✔ purrr   1.0.1
## ✔ tibble  3.2.1     ✔ dplyr   1.1.2
## ✔ tidyr   1.3.0     ✔ stringr 1.5.0
## ✔ readr   2.1.3     ✔ forcats 1.0.0
## ── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
## ✖ dplyr::filter() masks stats::filter()
## ✖ dplyr::lag()    masks stats::lag()</code></pre>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">=</span> out_df <span class="sc">%&gt;%</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Gene) <span class="sc">%&gt;%</span></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Screen.P =</span> isotwas<span class="sc">::</span><span class="fu">p_screen</span>(P))</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">=</span> <span class="fu">as.data.frame</span>(gene)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(gene)</span></code></pre></div>
<pre><code>##              Gene     Screen.P
## 1 ENSG00000070831 0.0016636386
## 2 ENSG00000162552 0.0008302377
## 3 ENSG00000230068 0.0004539979</code></pre>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a>alpha1<span class="ot">=</span>.<span class="dv">05</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a>G <span class="ot">=</span> <span class="fu">nrow</span>(gene)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>gene<span class="sc">$</span>Screen.P.Adjusted <span class="ot">=</span> <span class="fu">p.adjust</span>(gene<span class="sc">$</span>Screen.P,<span class="at">method =</span> <span class="st">&#39;fdr&#39;</span>)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>R <span class="ot">=</span> <span class="fu">length</span>(<span class="fu">unique</span>(gene<span class="sc">$</span>Gene[gene<span class="sc">$</span>Screen.P.Adjusted <span class="sc">&lt;</span> alpha1]))</span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>alpha2 <span class="ot">=</span> (R<span class="sc">*</span>alpha1)<span class="sc">/</span>G</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(gene)</span></code></pre></div>
<pre><code>##              Gene     Screen.P Screen.P.Adjusted
## 1 ENSG00000070831 0.0016636386       0.001663639
## 2 ENSG00000162552 0.0008302377       0.001245357
## 3 ENSG00000230068 0.0004539979       0.001245357</code></pre>
<p>We see that, on the gene-level, all 3 genes pass FDR control. That
is, we see that the adjusted screening P-value is less than 0.05 for all
3 genes. Now, we have to drop down to the isoform level for all 3
genes.</p>
<p>Let’s run the second stage: isoform-level FWER control for these gene
families.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="dv">0</span>,</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>                                   <span class="at">ncol =</span> <span class="fu">ncol</span>(out_df)<span class="sc">+</span><span class="dv">2</span>))</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(isoform_new) <span class="ot">=</span> <span class="fu">c</span>(<span class="fu">colnames</span>(out_df),<span class="st">&#39;Screen.P&#39;</span>,<span class="st">&#39;Confirmation.P&#39;</span>)</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">=</span> gene[<span class="fu">order</span>(gene<span class="sc">$</span>Screen.P),]</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true" tabindex="-1"></a>ttt <span class="ot">=</span> <span class="fu">merge</span>(out_df,</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true" tabindex="-1"></a>            gene[,<span class="fu">c</span>(<span class="st">&#39;Gene&#39;</span>,<span class="st">&#39;Screen.P&#39;</span>,</span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true" tabindex="-1"></a>                    <span class="st">&#39;Screen.P.Adjusted&#39;</span>)],</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true" tabindex="-1"></a>            <span class="at">by =</span> <span class="st">&#39;Gene&#39;</span>)</span>
<span id="cb16-9"><a href="#cb16-9" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb16-10"><a href="#cb16-10" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> ttt <span class="sc">%&gt;%</span></span>
<span id="cb16-11"><a href="#cb16-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">group_by</span>(Gene) <span class="sc">%&gt;%</span></span>
<span id="cb16-12"><a href="#cb16-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">summarise</span>(<span class="at">Feature =</span> Feature,</span>
<span id="cb16-13"><a href="#cb16-13" aria-hidden="true" tabindex="-1"></a>            <span class="at">Confirmation.P =</span> isotwas<span class="sc">::</span><span class="fu">p_confirm</span>(P,<span class="at">alpha =</span> alpha2))</span></code></pre></div>
<pre><code>## `summarise()` has grouped output by &#39;Gene&#39;. You can override using the
## `.groups` argument.</code></pre>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> <span class="fu">merge</span>(isoform_new,ttt,<span class="at">by=</span><span class="fu">c</span>(<span class="st">&#39;Gene&#39;</span>,<span class="st">&#39;Feature&#39;</span>))</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true" tabindex="-1"></a>isoform_new<span class="sc">$</span>Confirmation.P <span class="ot">=</span> <span class="fu">ifelse</span>(isoform_new<span class="sc">$</span>Screen.P.Adjusted <span class="sc">&lt;</span> <span class="fl">0.05</span>,</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true" tabindex="-1"></a>                                    isoform_new<span class="sc">$</span>Confirmation.P,</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true" tabindex="-1"></a>                                    <span class="dv">1</span>)</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> isoform_new[,<span class="fu">c</span>(<span class="st">&#39;Gene&#39;</span>,<span class="st">&#39;Feature&#39;</span>,<span class="st">&#39;Z&#39;</span>,<span class="st">&#39;P&#39;</span>,<span class="st">&#39;permute.P&#39;</span>,</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true" tabindex="-1"></a>                             <span class="st">&#39;topSNP&#39;</span>,<span class="st">&#39;topSNP.P&#39;</span>,</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true" tabindex="-1"></a>                             <span class="st">&#39;Screen.P&#39;</span>,<span class="st">&#39;Screen.P.Adjusted&#39;</span>,<span class="st">&#39;Confirmation.P&#39;</span>)]</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(isoform_new)</span></code></pre></div>
<pre><code>##               Gene         Feature         Z            P   permute.P
## 1  ENSG00000070831 ENST00000315554  1.268410 0.2046514827 1.000000000
## 2  ENSG00000070831 ENST00000344548 -3.239210 0.0011986108 1.000000000
## 3  ENSG00000070831 ENST00000400259  3.344840 0.0008232993 0.010989011
## 4  ENSG00000070831 ENST00000498236 -2.906460 0.0036554386 1.000000000
## 5  ENSG00000070831 ENST00000656825  3.268385 0.0010816323 1.000000000
## 6  ENSG00000070831 ENST00000662562 -3.278162 0.0010448529 1.000000000
## 7  ENSG00000070831 ENST00000667384  0.575199 0.5651567515 1.000000000
## 8  ENSG00000162552 ENST00000290167 -3.441877 0.0005776926 0.047952048
## 9  ENSG00000162552 ENST00000415567 -3.179540 0.0014750883 1.000000000
## 10 ENSG00000230068 ENST00000431803 -3.506525 0.0004539979 0.005994006
##        topSNP     topSNP.P     Screen.P Screen.P.Adjusted Confirmation.P
## 1  rs11582542 0.0003396706 0.0016636386       0.001663639   1.0000000000
## 2   rs2143103 0.0004331767 0.0016636386       0.001663639   0.0062691173
## 3  rs11582542 0.0003396706 0.0016636386       0.001663639   0.0049397958
## 4  rs11582542 0.0003396706 0.0016636386       0.001663639   0.0109663157
## 5  rs11582542 0.0003396706 0.0016636386       0.001663639   0.0062691173
## 6   rs2143103 0.0004331767 0.0016636386       0.001663639   0.0062691173
## 7  rs11582542 0.0003396706 0.0016636386       0.001663639   1.0000000000
## 8  rs11582542 0.0003396706 0.0008302377       0.001245357   0.0005776926
## 9  rs11582542 0.0003396706 0.0008302377       0.001245357   0.0014750883
## 10 rs11582542 0.0003396706 0.0004539979       0.001245357   0.0000000000</code></pre>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(<span class="fu">subset</span>(isoform_new,Screen.P.Adjusted <span class="sc">&lt;</span> alpha1 <span class="sc">&amp;</span></span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true" tabindex="-1"></a>               Confirmation.P <span class="sc">&lt;</span> alpha2 <span class="sc">&amp;</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true" tabindex="-1"></a>               permute.P <span class="sc">&lt;</span> <span class="fl">0.05</span>))</span></code></pre></div>
<pre><code>##               Gene         Feature         Z            P   permute.P
## 3  ENSG00000070831 ENST00000400259  3.344840 0.0008232993 0.010989011
## 8  ENSG00000162552 ENST00000290167 -3.441877 0.0005776926 0.047952048
## 10 ENSG00000230068 ENST00000431803 -3.506525 0.0004539979 0.005994006
##        topSNP     topSNP.P     Screen.P Screen.P.Adjusted Confirmation.P
## 3  rs11582542 0.0003396706 0.0016636386       0.001663639   0.0049397958
## 8  rs11582542 0.0003396706 0.0008302377       0.001245357   0.0005776926
## 10 rs11582542 0.0003396706 0.0004539979       0.001245357   0.0000000000</code></pre>
<p>We have now generated screening gene-level FDR-adjusted P-values and
confirmation isoform-level FWER controlled P-values. We see that there
are 3 isoforms at this locus that pass all 3 levels of hypothesis
testing: screening FDR-adjusted P &lt; 0.05, confirmation FWER-adjusted
P &lt; 0.05, and nominal permutation P &lt; 0.05. These isoforms are
ENST00000400259 (gene: <em>ENSG00000070831</em>), ENST00000290167 (gene:
<em>ENSG00000162552</em>), and ENST00000431803 (gene:
<em>ENSG00000230068</em>).</p>
</div>
<div id="fine-mapping-overlapping-and-trait-associated-isoforms" class="section level2">
<h2>Fine-mapping overlapping and trait-associated isoforms</h2>
<p>Since these two isoforms overlap and likely share LD-linked SNPs that
predict each isoform, we can’t be sure which of the three isoforms are
likely to be carrying the trait association signal. In short, we account
of the correlations between the genetically-predicted portions of the
expression of these isoforms, enumerate the possible causal
configurations, and use a Bayesian approach to generate a posterior
inclusion probability (PIP) for each isoform. These PIPs can be used to
generate a 90% credible set of isoforms that best explain the trait
association at the locus. See <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6619422/">Mancuso et
al’s manuscript</a> about gene-level fine-mapping for more of the
intuition and mathematical details.</p>
<p>For fine-mapping, you no longer need the GWAS summary statistics. You
will need:</p>
<ol style="list-style-type: decimal">
<li><p>TWAS summary statistics (something that looks like
<code>isoform_new</code>),</p></li>
<li><p>the isoTWAS models, and</p></li>
<li><p>an LD reference panel.</p></li>
</ol>
<p>Here, you need to make sure the LD reference matrix contains
<strong>all SNPs that are in all models for all isoforms in the
overlapping locus</strong>. If you do not have the in-sample
individual-level genotypes, we recommend using 1000 Genomes. For ease in
this tutorial, we have compiled the necessary LD matrix.</p>
<p>First, we annotate each isoform with location and select the
significantly associated isoforms that are within 1 Mb of one
another.</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="fu">suppressPackageStartupMessages</span>(<span class="fu">require</span>(biomaRt))</span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a>gene_names <span class="ot">=</span> <span class="fu">unique</span>(isoform_new<span class="sc">$</span>Gene)</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>ensembl <span class="ot">&lt;-</span> <span class="fu">useEnsembl</span>(<span class="at">biomart =</span> <span class="st">&quot;ensembl&quot;</span>, </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>                   <span class="at">dataset =</span> <span class="st">&quot;hsapiens_gene_ensembl&quot;</span>, </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>                   <span class="at">mirror =</span> <span class="st">&quot;useast&quot;</span>)</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>bm <span class="ot">=</span> <span class="fu">getBM</span>(<span class="at">attributes =</span> <span class="fu">c</span>(<span class="st">&#39;ensembl_gene_id&#39;</span>, </span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>                          <span class="st">&#39;chromosome_name&#39;</span>,</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>                          <span class="st">&#39;start_position&#39;</span>,</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>                          <span class="st">&#39;end_position&#39;</span>),</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true" tabindex="-1"></a>      <span class="at">filters =</span> <span class="st">&#39;ensembl_gene_id&#39;</span>,</span>
<span id="cb22-11"><a href="#cb22-11" aria-hidden="true" tabindex="-1"></a>      <span class="at">values =</span> gene_names, </span>
<span id="cb22-12"><a href="#cb22-12" aria-hidden="true" tabindex="-1"></a>      <span class="at">mart =</span> ensembl)</span>
<span id="cb22-13"><a href="#cb22-13" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(bm) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;Gene&#39;</span>,<span class="st">&#39;Chromosome&#39;</span>,<span class="st">&#39;Start&#39;</span>,<span class="st">&#39;End&#39;</span>)</span>
<span id="cb22-14"><a href="#cb22-14" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> <span class="fu">merge</span>(bm,isoform_new,<span class="at">by=</span><span class="st">&#39;Gene&#39;</span>)</span>
<span id="cb22-15"><a href="#cb22-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-16"><a href="#cb22-16" aria-hidden="true" tabindex="-1"></a>isoform_new <span class="ot">=</span> isoform_new[<span class="fu">order</span>(isoform_new<span class="sc">$</span>Chromosome,</span>
<span id="cb22-17"><a href="#cb22-17" aria-hidden="true" tabindex="-1"></a>                                isoform_new<span class="sc">$</span>Start,</span>
<span id="cb22-18"><a href="#cb22-18" aria-hidden="true" tabindex="-1"></a>                                <span class="at">decreasing =</span> F),]</span>
<span id="cb22-19"><a href="#cb22-19" aria-hidden="true" tabindex="-1"></a>isoform_sig <span class="ot">=</span> <span class="fu">subset</span>(isoform_new,</span>
<span id="cb22-20"><a href="#cb22-20" aria-hidden="true" tabindex="-1"></a>                     Screen.P.Adjusted <span class="sc">&lt;</span> alpha1 <span class="sc">&amp;</span></span>
<span id="cb22-21"><a href="#cb22-21" aria-hidden="true" tabindex="-1"></a>                       Confirmation.P <span class="sc">&lt;</span> alpha2 <span class="sc">&amp;</span></span>
<span id="cb22-22"><a href="#cb22-22" aria-hidden="true" tabindex="-1"></a>                       permute.P <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb22-23"><a href="#cb22-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb22-24"><a href="#cb22-24" aria-hidden="true" tabindex="-1"></a>keep.isoform <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb22-25"><a href="#cb22-25" aria-hidden="true" tabindex="-1"></a><span class="cf">if</span> (<span class="fu">nrow</span>(isoform_sig) <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb22-26"><a href="#cb22-26" aria-hidden="true" tabindex="-1"></a>      <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span>(<span class="fu">nrow</span>(isoform_sig)<span class="sc">-</span><span class="dv">1</span>)){</span>
<span id="cb22-27"><a href="#cb22-27" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> (isoform_sig<span class="sc">$</span>End[i] <span class="sc">&gt;</span> isoform_sig<span class="sc">$</span>Start[i<span class="sc">+</span><span class="dv">1</span>] <span class="sc">-</span> <span class="fl">1e6</span>){</span>
<span id="cb22-28"><a href="#cb22-28" aria-hidden="true" tabindex="-1"></a>          keep.isoform <span class="ot">=</span> <span class="fu">unique</span>(<span class="fu">c</span>(keep.isoform,</span>
<span id="cb22-29"><a href="#cb22-29" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">c</span>(isoform_sig<span class="sc">$</span>Feature[<span class="fu">c</span>(i,i<span class="sc">+</span><span class="dv">1</span>)])))</span>
<span id="cb22-30"><a href="#cb22-30" aria-hidden="true" tabindex="-1"></a>        }</span>
<span id="cb22-31"><a href="#cb22-31" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb22-32"><a href="#cb22-32" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb22-33"><a href="#cb22-33" aria-hidden="true" tabindex="-1"></a>isoform_sig <span class="ot">=</span> <span class="fu">subset</span>(isoform_sig,Feature <span class="sc">%in%</span> keep.isoform)</span></code></pre></div>
<p>Significant isoforms that do not overlap within 1 Mb of another
significant isoform do not need to be fine-mapped.</p>
<p>Now, we aggregate the models and generate a single table with the
SNP-to-isoform weights of all the SNPs that predict these overlapping
isoforms.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true" tabindex="-1"></a>all.snps <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true" tabindex="-1"></a>omega <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true" tabindex="-1"></a>pos <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb23-4"><a href="#cb23-4" aria-hidden="true" tabindex="-1"></a>gene <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb23-5"><a href="#cb23-5" aria-hidden="true" tabindex="-1"></a>snp.chr <span class="ot">=</span> <span class="fu">c</span>()</span>
<span id="cb23-6"><a href="#cb23-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-7"><a href="#cb23-7" aria-hidden="true" tabindex="-1"></a><span class="do">### COLLECT WEIGHTS FOR SNPS IN THE MODELS</span></span>
<span id="cb23-8"><a href="#cb23-8" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(isoform_sig)){</span>
<span id="cb23-9"><a href="#cb23-9" aria-hidden="true" tabindex="-1"></a>  gene_in <span class="ot">=</span> isoform_sig<span class="sc">$</span>Gene[i]</span>
<span id="cb23-10"><a href="#cb23-10" aria-hidden="true" tabindex="-1"></a>  model_in <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb23-11"><a href="#cb23-11" aria-hidden="true" tabindex="-1"></a>                              <span class="fu">paste0</span>(gene_in,<span class="st">&quot;_isoTWAS.RDS&quot;</span>),</span>
<span id="cb23-12"><a href="#cb23-12" aria-hidden="true" tabindex="-1"></a>                              <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span>
<span id="cb23-13"><a href="#cb23-13" aria-hidden="true" tabindex="-1"></a>  model_in <span class="ot">=</span> <span class="fu">subset</span>(model_in,</span>
<span id="cb23-14"><a href="#cb23-14" aria-hidden="true" tabindex="-1"></a>                    Feature <span class="sc">==</span> isoform_sig<span class="sc">$</span>Feature[i])</span>
<span id="cb23-15"><a href="#cb23-15" aria-hidden="true" tabindex="-1"></a>  Model <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">SNP =</span> model_in<span class="sc">$</span>SNP,</span>
<span id="cb23-16"><a href="#cb23-16" aria-hidden="true" tabindex="-1"></a>                     <span class="at">Chromosome =</span> model_in<span class="sc">$</span>Chromosome,</span>
<span id="cb23-17"><a href="#cb23-17" aria-hidden="true" tabindex="-1"></a>                     <span class="at">Position =</span> model_in<span class="sc">$</span>Position,</span>
<span id="cb23-18"><a href="#cb23-18" aria-hidden="true" tabindex="-1"></a>                     <span class="at">Effect =</span> model_in<span class="sc">$</span>Weight,</span>
<span id="cb23-19"><a href="#cb23-19" aria-hidden="true" tabindex="-1"></a>                     <span class="at">A1 =</span> model_in<span class="sc">$</span>ALT,</span>
<span id="cb23-20"><a href="#cb23-20" aria-hidden="true" tabindex="-1"></a>                     <span class="at">A2 =</span> model_in<span class="sc">$</span>REF)</span>
<span id="cb23-21"><a href="#cb23-21" aria-hidden="true" tabindex="-1"></a>  Model <span class="ot">=</span> <span class="fu">subset</span>(Model,Effect<span class="sc">!=</span><span class="dv">0</span>)</span>
<span id="cb23-22"><a href="#cb23-22" aria-hidden="true" tabindex="-1"></a>  Model <span class="ot">=</span> Model[<span class="sc">!</span><span class="fu">duplicated</span>(Model<span class="sc">$</span>SNP),]</span>
<span id="cb23-23"><a href="#cb23-23" aria-hidden="true" tabindex="-1"></a>  all.snps <span class="ot">=</span> <span class="fu">c</span>(all.snps,</span>
<span id="cb23-24"><a href="#cb23-24" aria-hidden="true" tabindex="-1"></a>               <span class="fu">as.character</span>(Model<span class="sc">$</span>SNP))</span>
<span id="cb23-25"><a href="#cb23-25" aria-hidden="true" tabindex="-1"></a>  omega <span class="ot">=</span> <span class="fu">c</span>(omega,</span>
<span id="cb23-26"><a href="#cb23-26" aria-hidden="true" tabindex="-1"></a>            <span class="fu">as.numeric</span>(Model<span class="sc">$</span>Effect))</span>
<span id="cb23-27"><a href="#cb23-27" aria-hidden="true" tabindex="-1"></a>  gene <span class="ot">=</span> <span class="fu">c</span>(gene,</span>
<span id="cb23-28"><a href="#cb23-28" aria-hidden="true" tabindex="-1"></a>           <span class="fu">rep</span>(isoform_sig<span class="sc">$</span>Feature[i],<span class="fu">nrow</span>(Model)))</span>
<span id="cb23-29"><a href="#cb23-29" aria-hidden="true" tabindex="-1"></a>  snp.chr <span class="ot">=</span> <span class="fu">c</span>(snp.chr,</span>
<span id="cb23-30"><a href="#cb23-30" aria-hidden="true" tabindex="-1"></a>              <span class="fu">as.numeric</span>(Model<span class="sc">$</span>Chromosome))</span>
<span id="cb23-31"><a href="#cb23-31" aria-hidden="true" tabindex="-1"></a>  pos <span class="ot">=</span> <span class="fu">c</span>(pos,<span class="fu">as.numeric</span>(Model<span class="sc">$</span>Position))</span>
<span id="cb23-32"><a href="#cb23-32" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-33"><a href="#cb23-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-34"><a href="#cb23-34" aria-hidden="true" tabindex="-1"></a>tot.df <span class="ot">=</span> <span class="fu">data.frame</span>(<span class="at">SNP =</span> all.snps,</span>
<span id="cb23-35"><a href="#cb23-35" aria-hidden="true" tabindex="-1"></a>                    <span class="at">Gene =</span> gene,</span>
<span id="cb23-36"><a href="#cb23-36" aria-hidden="true" tabindex="-1"></a>                    <span class="at">Effect =</span> omega,</span>
<span id="cb23-37"><a href="#cb23-37" aria-hidden="true" tabindex="-1"></a>                    <span class="at">Chromosome =</span> snp.chr)</span>
<span id="cb23-38"><a href="#cb23-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-39"><a href="#cb23-39" aria-hidden="true" tabindex="-1"></a>model.df <span class="ot">=</span> <span class="fu">as.data.frame</span>(<span class="fu">matrix</span>(<span class="at">nrow =</span> <span class="fu">length</span>(<span class="fu">unique</span>(all.snps)),</span>
<span id="cb23-40"><a href="#cb23-40" aria-hidden="true" tabindex="-1"></a>                                <span class="at">ncol =</span> <span class="fu">nrow</span>(isoform_sig)<span class="sc">+</span><span class="dv">1</span>))</span>
<span id="cb23-41"><a href="#cb23-41" aria-hidden="true" tabindex="-1"></a><span class="fu">colnames</span>(model.df) <span class="ot">=</span> <span class="fu">c</span>(<span class="st">&#39;SNP&#39;</span>,isoform_sig<span class="sc">$</span>Feature)</span>
<span id="cb23-42"><a href="#cb23-42" aria-hidden="true" tabindex="-1"></a>model.df<span class="sc">$</span>SNP <span class="ot">=</span> <span class="fu">as.character</span>(<span class="fu">unique</span>(all.snps))</span>
<span id="cb23-43"><a href="#cb23-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-44"><a href="#cb23-44" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (q <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(isoform_sig)){</span>
<span id="cb23-45"><a href="#cb23-45" aria-hidden="true" tabindex="-1"></a>  cur.tot.df <span class="ot">=</span> <span class="fu">subset</span>(tot.df,Gene <span class="sc">==</span> isoform_sig<span class="sc">$</span>Feature[q])</span>
<span id="cb23-46"><a href="#cb23-46" aria-hidden="true" tabindex="-1"></a>  cur.tot.df<span class="sc">$</span>SNP <span class="ot">=</span> <span class="fu">as.character</span>(cur.tot.df<span class="sc">$</span>SNP)</span>
<span id="cb23-47"><a href="#cb23-47" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(model.df)){</span>
<span id="cb23-48"><a href="#cb23-48" aria-hidden="true" tabindex="-1"></a>    w <span class="ot">=</span> <span class="fu">which</span>(cur.tot.df<span class="sc">$</span>SNP <span class="sc">==</span> model.df<span class="sc">$</span>SNP[i])</span>
<span id="cb23-49"><a href="#cb23-49" aria-hidden="true" tabindex="-1"></a>    model.df[i,q<span class="sc">+</span><span class="dv">1</span>] <span class="ot">=</span> <span class="fu">ifelse</span>(<span class="fu">length</span>(w) <span class="sc">!=</span> <span class="dv">0</span>,</span>
<span id="cb23-50"><a href="#cb23-50" aria-hidden="true" tabindex="-1"></a>                             cur.tot.df<span class="sc">$</span>Effect[w],</span>
<span id="cb23-51"><a href="#cb23-51" aria-hidden="true" tabindex="-1"></a>                             <span class="dv">0</span>)</span>
<span id="cb23-52"><a href="#cb23-52" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb23-53"><a href="#cb23-53" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-54"><a href="#cb23-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-55"><a href="#cb23-55" aria-hidden="true" tabindex="-1"></a>model.df<span class="sc">$</span>Chromosome <span class="ot">=</span> <span class="dv">2</span></span>
<span id="cb23-56"><a href="#cb23-56" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(model.df)){</span>
<span id="cb23-57"><a href="#cb23-57" aria-hidden="true" tabindex="-1"></a>  rrr <span class="ot">=</span> <span class="fu">subset</span>(tot.df,SNP <span class="sc">==</span> model.df<span class="sc">$</span>SNP[i])</span>
<span id="cb23-58"><a href="#cb23-58" aria-hidden="true" tabindex="-1"></a>  model.df<span class="sc">$</span>Chromosome[i] <span class="ot">=</span> rrr<span class="sc">$</span>Chromosome[<span class="dv">1</span>]</span>
<span id="cb23-59"><a href="#cb23-59" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb23-60"><a href="#cb23-60" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb23-61"><a href="#cb23-61" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(model.df)</span></code></pre></div>
<pre><code>##          SNP ENST00000400259 ENST00000431803 ENST00000290167 Chromosome
## 1  rs1002480   -3.437902e-04    8.726253e-05               0          1
## 2  rs1011380   -5.231286e-05    8.881914e-05               0          1
## 3  rs1014986    7.976357e-05    1.177811e-03               0          1
## 4 rs10157808    2.620658e-04    2.047407e-05               0          1
## 5  rs1018102    4.699228e-05    1.019356e-04               0          1
## 6 rs10218584   -6.270159e-05    3.580614e-06               0          1</code></pre>
<p>This step generates a data frame in <code>model.df</code> that
includes all SNPs in both isoTWAS models at the locus and their effect
(non-zero or otherwise) on each isoforms we are fine-mapping at the
locus.</p>
<p>Now, we obtain our LD matrix for the SNPs in
<code>model.df$SNP</code></p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>V <span class="ot">=</span> <span class="fu">readRDS</span>(<span class="fu">system.file</span>(<span class="st">&quot;extdata&quot;</span>,</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>                        <span class="st">&quot;test_LD_finemapping.RDS&quot;</span>,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>                        <span class="at">package =</span> <span class="st">&quot;isotwas&quot;</span>))</span></code></pre></div>
<p>Now, we run the fine-mapping process in the following steps: (1)
compute the correlation between the genetically-predicted expression of
the isoforms, (2) enumerate the causal configurations, (3) compute
marginal likelihoods and Bayes factors for each isoform, and (4) compute
posterior inclusion probabilities and credible sets.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true" tabindex="-1"></a>V <span class="ot">=</span> V[model.df<span class="sc">$</span>SNP,model.df<span class="sc">$</span>SNP]</span>
<span id="cb26-2"><a href="#cb26-2" aria-hidden="true" tabindex="-1"></a>Omega <span class="ot">=</span> Matrix<span class="sc">::</span><span class="fu">Matrix</span>(<span class="fu">as.matrix</span>(model.df[,<span class="sc">-</span><span class="fu">c</span>(<span class="dv">1</span>,<span class="fu">ncol</span>(model.df))]))</span>
<span id="cb26-3"><a href="#cb26-3" aria-hidden="true" tabindex="-1"></a>zscores <span class="ot">=</span> isoform_sig<span class="sc">$</span>Z</span>
<span id="cb26-4"><a href="#cb26-4" aria-hidden="true" tabindex="-1"></a>m <span class="ot">=</span> <span class="fu">length</span>(zscores)</span>
<span id="cb26-5"><a href="#cb26-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-6"><a href="#cb26-6" aria-hidden="true" tabindex="-1"></a><span class="do">### COMPUTE LD BETWEEN TX ON THE GENETIC LEVEL</span></span>
<span id="cb26-7"><a href="#cb26-7" aria-hidden="true" tabindex="-1"></a>wcor <span class="ot">=</span> isotwas<span class="sc">::</span><span class="fu">estimate_cor</span>(<span class="fu">as.matrix</span>(Omega),</span>
<span id="cb26-8"><a href="#cb26-8" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">as.matrix</span>(V),</span>
<span id="cb26-9"><a href="#cb26-9" aria-hidden="true" tabindex="-1"></a>                             <span class="at">intercept=</span>T)[[<span class="dv">1</span>]]</span>
<span id="cb26-10"><a href="#cb26-10" aria-hidden="true" tabindex="-1"></a><span class="fu">diag</span>(wcor) <span class="ot">=</span> <span class="dv">1</span></span>
<span id="cb26-11"><a href="#cb26-11" aria-hidden="true" tabindex="-1"></a>wcor[<span class="fu">is.na</span>(wcor)] <span class="ot">=</span> <span class="dv">0</span></span>
<span id="cb26-12"><a href="#cb26-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-13"><a href="#cb26-13" aria-hidden="true" tabindex="-1"></a><span class="do">### COMPUTE LD INTERCEPT BETWEEN ISOFRM ON THE GENETIC LEVEL</span></span>
<span id="cb26-14"><a href="#cb26-14" aria-hidden="true" tabindex="-1"></a>swld <span class="ot">=</span> isotwas<span class="sc">::</span><span class="fu">estimate_cor</span>(<span class="fu">as.matrix</span>(Omega),</span>
<span id="cb26-15"><a href="#cb26-15" aria-hidden="true" tabindex="-1"></a>                             <span class="fu">as.matrix</span>(V),</span>
<span id="cb26-16"><a href="#cb26-16" aria-hidden="true" tabindex="-1"></a>                             <span class="at">intercept=</span>T)[[<span class="dv">2</span>]]</span>
<span id="cb26-17"><a href="#cb26-17" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb26-18"><a href="#cb26-18" aria-hidden="true" tabindex="-1"></a>null_res <span class="ot">=</span> m <span class="sc">*</span> <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fl">1e-3</span>)</span>
<span id="cb26-19"><a href="#cb26-19" aria-hidden="true" tabindex="-1"></a>marginal <span class="ot">=</span> m <span class="sc">*</span> <span class="fu">log</span>(<span class="dv">1</span> <span class="sc">-</span> <span class="fl">1e-3</span>)</span>
<span id="cb26-20"><a href="#cb26-20" aria-hidden="true" tabindex="-1"></a>comb_list <span class="ot">=</span> <span class="fu">list</span>()</span>
<span id="cb26-21"><a href="#cb26-21" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (n <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">min</span>(<span class="dv">3</span>,<span class="fu">length</span>(zscores))){</span>
<span id="cb26-22"><a href="#cb26-22" aria-hidden="true" tabindex="-1"></a>  comb_list <span class="ot">=</span> <span class="fu">c</span>(comb_list,</span>
<span id="cb26-23"><a href="#cb26-23" aria-hidden="true" tabindex="-1"></a>                <span class="fu">combn</span>(<span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(zscores),n,<span class="at">simplify=</span>F))</span>
<span id="cb26-24"><a href="#cb26-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-25"><a href="#cb26-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-26"><a href="#cb26-26" aria-hidden="true" tabindex="-1"></a>pips <span class="ot">=</span> <span class="fu">rep</span>(<span class="dv">0</span>,<span class="fu">length</span>(zscores))</span>
<span id="cb26-27"><a href="#cb26-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-28"><a href="#cb26-28" aria-hidden="true" tabindex="-1"></a><span class="do">### COMPUTE BAYES FACTORS/LIKELIHOOD AT EACH CAUSAL CONFIG</span></span>
<span id="cb26-29"><a href="#cb26-29" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(comb_list)){</span>
<span id="cb26-30"><a href="#cb26-30" aria-hidden="true" tabindex="-1"></a>  subset <span class="ot">=</span> comb_list[[j]]</span>
<span id="cb26-31"><a href="#cb26-31" aria-hidden="true" tabindex="-1"></a>  local <span class="ot">=</span> isotwas<span class="sc">::</span><span class="fu">bayes_factor</span>(zscores,</span>
<span id="cb26-32"><a href="#cb26-32" aria-hidden="true" tabindex="-1"></a>                                <span class="at">idx_set =</span> subset,</span>
<span id="cb26-33"><a href="#cb26-33" aria-hidden="true" tabindex="-1"></a>                                <span class="at">wcor =</span> wcor)</span>
<span id="cb26-34"><a href="#cb26-34" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb26-35"><a href="#cb26-35" aria-hidden="true" tabindex="-1"></a>  marginal <span class="ot">=</span> <span class="fu">log</span>(<span class="fu">exp</span>(local) <span class="sc">+</span> <span class="fu">exp</span>(marginal))</span>
<span id="cb26-36"><a href="#cb26-36" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span> (idx <span class="cf">in</span> subset){</span>
<span id="cb26-37"><a href="#cb26-37" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (pips[idx] <span class="sc">==</span> <span class="dv">0</span>){</span>
<span id="cb26-38"><a href="#cb26-38" aria-hidden="true" tabindex="-1"></a>      pips[idx] <span class="ot">=</span> local</span>
<span id="cb26-39"><a href="#cb26-39" aria-hidden="true" tabindex="-1"></a>      } <span class="cf">else</span> {</span>
<span id="cb26-40"><a href="#cb26-40" aria-hidden="true" tabindex="-1"></a>        pips[idx] <span class="ot">=</span> <span class="fu">log</span>(<span class="fu">exp</span>(pips[idx]) <span class="sc">+</span> <span class="fu">exp</span>(local))</span>
<span id="cb26-41"><a href="#cb26-41" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-42"><a href="#cb26-42" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-43"><a href="#cb26-43" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-44"><a href="#cb26-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-45"><a href="#cb26-45" aria-hidden="true" tabindex="-1"></a>pips <span class="ot">=</span> <span class="fu">exp</span>(pips <span class="sc">-</span> marginal)</span>
<span id="cb26-46"><a href="#cb26-46" aria-hidden="true" tabindex="-1"></a>null_res <span class="ot">=</span> <span class="fu">exp</span>(null_res <span class="sc">-</span> marginal)</span>
<span id="cb26-47"><a href="#cb26-47" aria-hidden="true" tabindex="-1"></a>isoform_sig<span class="sc">$</span>pip <span class="ot">=</span> pips</span>
<span id="cb26-48"><a href="#cb26-48" aria-hidden="true" tabindex="-1"></a>isoform_sig <span class="ot">=</span> isoform_sig[<span class="fu">order</span>(isoform_sig<span class="sc">$</span>pip,<span class="at">decreasing =</span> T),]</span>
<span id="cb26-49"><a href="#cb26-49" aria-hidden="true" tabindex="-1"></a>npost <span class="ot">=</span> isoform_sig<span class="sc">$</span>pip<span class="sc">/</span><span class="fu">sum</span>(isoform_sig<span class="sc">$</span>pip)</span>
<span id="cb26-50"><a href="#cb26-50" aria-hidden="true" tabindex="-1"></a>csum <span class="ot">=</span> <span class="fu">cumsum</span>(npost)</span>
<span id="cb26-51"><a href="#cb26-51" aria-hidden="true" tabindex="-1"></a>isoform_sig<span class="sc">$</span>in_cred_set <span class="ot">=</span> F</span>
<span id="cb26-52"><a href="#cb26-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-53"><a href="#cb26-53" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(isoform_sig)){</span>
<span id="cb26-54"><a href="#cb26-54" aria-hidden="true" tabindex="-1"></a>  isoform_sig<span class="sc">$</span>in_cred_set[i] <span class="ot">=</span> T</span>
<span id="cb26-55"><a href="#cb26-55" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span> (i <span class="sc">&gt;</span> <span class="dv">1</span>){</span>
<span id="cb26-56"><a href="#cb26-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (csum[i] <span class="sc">&gt;</span> .<span class="dv">9</span> <span class="sc">&amp;</span> csum[i<span class="dv">-1</span>] <span class="sc">&lt;</span> .<span class="dv">9</span>){</span>
<span id="cb26-57"><a href="#cb26-57" aria-hidden="true" tabindex="-1"></a>      isoform_sig<span class="sc">$</span>in_cred_set[i] <span class="ot">=</span> T</span>
<span id="cb26-58"><a href="#cb26-58" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb26-59"><a href="#cb26-59" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (csum[i] <span class="sc">&lt;</span> .<span class="dv">9</span>){</span>
<span id="cb26-60"><a href="#cb26-60" aria-hidden="true" tabindex="-1"></a>      isoform_sig<span class="sc">$</span>in_cred_set[i] <span class="ot">=</span> T</span>
<span id="cb26-61"><a href="#cb26-61" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb26-62"><a href="#cb26-62" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> (csum[i] <span class="sc">&gt;</span> .<span class="dv">9</span> <span class="sc">&amp;</span> csum[i<span class="dv">-1</span>] <span class="sc">&gt;</span> .<span class="dv">9</span>){</span>
<span id="cb26-63"><a href="#cb26-63" aria-hidden="true" tabindex="-1"></a>      isoform_sig<span class="sc">$</span>in_cred_set[i] <span class="ot">=</span> F</span>
<span id="cb26-64"><a href="#cb26-64" aria-hidden="true" tabindex="-1"></a>      }</span>
<span id="cb26-65"><a href="#cb26-65" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb26-66"><a href="#cb26-66" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb26-67"><a href="#cb26-67" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb26-68"><a href="#cb26-68" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(isoform_sig)</span></code></pre></div>
<pre><code>##               Gene Chromosome    Start      End         Feature         Z
## 3  ENSG00000070831          1 22052627 22101360 ENST00000400259  3.344840
## 10 ENSG00000230068          1 22059197 22064199 ENST00000431803 -3.506525
## 8  ENSG00000162552          1 22117313 22143969 ENST00000290167 -3.441877
##               P   permute.P     topSNP     topSNP.P     Screen.P
## 3  0.0008232993 0.010989011 rs11582542 0.0003396706 0.0016636386
## 10 0.0004539979 0.005994006 rs11582542 0.0003396706 0.0004539979
## 8  0.0005776926 0.047952048 rs11582542 0.0003396706 0.0008302377
##    Screen.P.Adjusted Confirmation.P       pip in_cred_set
## 3        0.001663639   0.0049397958 1.0000000        TRUE
## 10       0.001245357   0.0000000000 1.0000000        TRUE
## 8        0.001245357   0.0005776926 0.2439366        TRUE</code></pre>
<p>As you can see, ENST00000431803 and ENST00000290167 have
<code>PIP = 1</code> and are both in the 90% credible set
(<code>in_cred_set</code> column).</p>
<p>Full scripts for association testing genome-wide are available <a href="https://github.com/bhattacharya-a-bt/isotwas_manu_scripts">here</a>.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
