<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Train an isoTWAS model • isotwas</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Train an isoTWAS model">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">isotwas</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">2.0.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/motivation.html">Motivation for isotwas</a></li>
    <li><a class="dropdown-item" href="../articles/pipeline.html">Train an isoTWAS model</a></li>
    <li><a class="dropdown-item" href="../articles/stageassoc.html">Run isoTWAS associations</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Train an isoTWAS model</h1>
            
      

      <div class="d-none name"><code>pipeline.Rmd</code></div>
    </div>

    
    
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>This module will cover how to use functions from <code>isotwas</code>
to train an multivariate <code>isotwas</code> model for isoform-level
expression. Along the way, we’ll be covering some practical
considerations for computation and feature selection.</p>
</div>
<div class="section level2">
<h2 id="train-predictive-model">Train predictive model<a class="anchor" aria-label="anchor" href="#train-predictive-model"></a>
</h2>
<p>To train a multivariate model, we need:</p>
<ol style="list-style-type: decimal">
<li><p>a matrix of dosages of SNPs, and</p></li>
<li><p>a matrix of isoform-level expressions for a given gene.</p></li>
</ol>
<p>Isoform expression estimated from pseudo-alignment methods like <a href="https://combine-lab.github.io/salmon/" class="external-link"><code>salmon</code></a> and
<a href="https://pachterlab.github.io/kallisto/" class="external-link"><code>kallisto</code></a>
provide bootstrapped estimates expression across a number of iterations.
This information can be included into model training, as illustrated
below.</p>
<p><u><strong>Practical consideration</strong></u>: Ideally, to reduce
your memory usage, we recommend storing your genetic data as
<code>PLINK</code> format data (ideally as
<code>.bed</code>/<code>bim</code>/<code>fam</code> or
<code>.pgen</code>/<code>.pvar</code>/<code>.psam</code>). For a given
gene, use <code>PLINK</code> to <a href="https://www.cog-genomics.org/plink/2.0/filter" class="external-link">select only
SNPs</a> on the same chromosome and within 1 Megabase of the gene body.
Then, you can use the <a href="https://privefl.github.io/bigsnpr/" class="external-link"><code>bigsnpr</code>
package</a> read in this “condensed” file set. Make sure your isoform
expression matrix is log-transformed and residualized by covariates that
are normally used in a QTL analysis (i.e., age, sex, principal
components of the genotype matrix, <a href="https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3398141/" class="external-link">PEER
factors</a> or <a href="https://pubmed.ncbi.nlm.nih.gov/23874524/" class="external-link">hidden covariates with
prior</a> for the full expression matrix). In order to ensure maximal
overlap between the model and the GWAS we will eventually use for trait
mapping, we suggest restricting the SNPs in the QTL panel to only those
present in the LD reference panel or to those annotated by the <a href="https://www.nature.com/articles/nature02168" class="external-link">HapMap3 Project</a>.
This will ensure proper standard error calculations.</p>
<p>For this vignette about training an model, we’ll use some toy genetic
and transcriptomic data provided.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">train_bed</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"train.bed"</span>, package <span class="op">=</span> <span class="st">"isotwas"</span><span class="op">)</span></span>
<span><span class="va">snps</span> <span class="op">=</span> <span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_attach.html" class="external-link">snp_attach</a></span><span class="op">(</span><span class="fu">bigsnpr</span><span class="fu">::</span><span class="fu"><a href="https://privefl.github.io/bigsnpr/reference/snp_readBed.html" class="external-link">snp_readBed2</a></span><span class="op">(</span><span class="va">train_bed</span>,</span>
<span>                                                 backingfile <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/tempfile.html" class="external-link">tempfile</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">snps</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1]  300 4542</span></span>
<span></span>
<span><span class="va">isoform_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"isoform_exp.RDS"</span>, package <span class="op">=</span> <span class="st">"isotwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">isoform_mat</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 300   3</span></span></code></pre></div>
<p>This reads in a SNP matrix with dimensions
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>300</mn><mo>×</mo><mn>4542</mn></mrow><annotation encoding="application/x-tex">300 \times 4542</annotation></semantics></math>,
with samples along the rows and SNPs along the columns and a
isoform-level expression matrix of dimension
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mn>300</mn><mo>×</mo><mn>3</mn></mrow><annotation encoding="application/x-tex">300 \times 3</annotation></semantics></math>,
with isoforms along the columns.</p>
<p><u><strong>Practical consideration</strong></u>: Though this is not a
requirement for the isotwas pipeline, we recommend a feature selection
step based on cis-heritability of isoform-level expression. Prior of
model training for isoforms of a given gene, we recommend that users
compute the heritability of all expressed isoforms of a gene using <a href="https://yanglab.westlake.edu.cn/software/gcta/#Overview" class="external-link"><code>GCTA</code></a>
(ideally the <a href="http://gusevlab.org/projects/fusion/" class="external-link"><code>gcta-nr-robust</code>
executable in <code>FUSION</code></a>). We recommend users only train
models for isoforms that have positive heritability at nominal
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>P</mi><mo>&lt;</mo><mn>0.05</mn></mrow><annotation encoding="application/x-tex">P &lt; 0.05</annotation></semantics></math>
or
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mn>0.01</mn><annotation encoding="application/x-tex">0.01</annotation></semantics></math>.</p>
<p><u><strong>Practical consideration</strong></u>: Prior to
heritability analysis and model training, it is imperative that your
expression data is residualized by relevant covariates (i.e. principal
components of genotype matrix, hidden covariates or PEER factors of the
expression matrix, clinical/demographic covariates). A general rule of
thumb is if you would include the covariate in an eQTL/isoform-level
eQTL analysis, you should include it in the residualization process.</p>
<p>Let’s jitter our expression matrix a little to generate 10
bootstrapped estimates of the matrix. Keep in mind that these
replicates, theoretically, should not affect the effect size estimates
but will help in seeding the estimation process of the correlation
between isoforms/error in the multivariate model.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">boot_list</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="kw">function</span><span class="op">(</span><span class="va">i</span><span class="op">)</span><span class="op">{</span><span class="fu"><a href="https://rdrr.io/r/base/jitter.html" class="external-link">jitter</a></span><span class="op">(</span><span class="va">isoform_mat</span><span class="op">)</span><span class="op">}</span><span class="op">)</span></span>
<span><span class="va">isoform_mat_rep</span> <span class="op">=</span> <span class="fu">rlist</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/rlist/man/list.rbind.html" class="external-link">list.rbind</a></span><span class="op">(</span><span class="va">boot_list</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Sample'</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">snps</span><span class="op">$</span><span class="va">fam</span><span class="op">)</span><span class="op">)</span>,<span class="fl">10</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">'Isoform'</span>,<span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/dim.html" class="external-link">dim</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 3000    3</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="available-prediction-methods">Available Prediction Methods<a class="anchor" aria-label="anchor" href="#available-prediction-methods"></a>
</h2>
<p>The <code><a href="../reference/compute_isotwas.html">compute_isotwas()</a></code> function supports 9 multivariate
prediction methods:</p>
<table class="table">
<colgroup>
<col width="38%">
<col width="61%">
</colgroup>
<thead><tr class="header">
<th>Method</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>mrce_lasso</code></td>
<td>Multivariate Regression with Covariance Estimation (MRCE) - jointly
estimates regression coefficients and error covariance</td>
</tr>
<tr class="even">
<td><code>multi_enet</code></td>
<td>Multivariate elastic net - fits elastic net for each isoform with
shared penalty</td>
</tr>
<tr class="odd">
<td><code>joinet</code></td>
<td>Joinet stacked elastic net - uses cross-predictions to improve
multivariate prediction</td>
</tr>
<tr class="even">
<td><code>spls</code></td>
<td>Sparse partial least squares - finds latent components that predict
all isoforms</td>
</tr>
<tr class="odd">
<td><code>sgl</code></td>
<td>Sparse group lasso - encourages both within-group and across-group
sparsity</td>
</tr>
<tr class="even">
<td><code>mtlasso</code></td>
<td>Multi-task lasso (L21) - joint feature selection across all
isoforms</td>
</tr>
<tr class="odd">
<td><code>stacking</code></td>
<td>Super learner stacking - combines predictions from multiple base
methods</td>
</tr>
<tr class="even">
<td><code>graph_reg</code></td>
<td>Graph-regularized regression - incorporates isoform similarity
structure</td>
</tr>
<tr class="odd">
<td><code>univariate</code></td>
<td>Best of univariate methods (elastic net, BLUP, SuSiE) per
isoform</td>
</tr>
</tbody>
</table>
</div>
<div class="section level2">
<h2 id="training-a-model">Training a Model<a class="anchor" aria-label="anchor" href="#training-a-model"></a>
</h2>
<p>Now, we can train the model using the <code><a href="../reference/compute_isotwas.html">compute_isotwas()</a></code>
function. Make sure the SNP matrix has clear labels for the SNP
identifiers that are relevant to your study.</p>
<p>The function now provides detailed progress output by default
(<code>verbose = TRUE</code>), showing:</p>
<ul>
<li>Data summary (samples, SNPs, transcripts)</li>
<li>Method-by-method progress with timing</li>
<li>Per-method R² performance</li>
<li>Final results comparison table</li>
</ul>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">snp_mat</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">snps</span><span class="op">$</span><span class="va">genotypes</span><span class="op">[</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">snp_mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">snps</span><span class="op">$</span><span class="va">map</span><span class="op">$</span><span class="va">marker.ID</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">snp_mat</span><span class="op">)</span> <span class="op">=</span> <span class="va">snps</span><span class="op">$</span><span class="va">fam</span><span class="op">$</span><span class="va">sample.ID</span></span>
<span></span>
<span><span class="co"># Train with all methods (recommended for best model selection)</span></span>
<span><span class="va">isotwas_model</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_isotwas.html">compute_isotwas</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">snp_mat</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">isoform_mat</span>,</span>
<span>  Y.rep <span class="op">=</span> <span class="va">isoform_mat_rep</span>,</span>
<span>  R <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span>,</span>
<span>  omega_est <span class="op">=</span> <span class="st">'replicates'</span>,</span>
<span>  omega_nlambda <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  nfolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1789</span>,</span>
<span>  run_all <span class="op">=</span> <span class="cn">TRUE</span>,        <span class="co"># Run all 9 methods</span></span>
<span>  return_all <span class="op">=</span> <span class="cn">TRUE</span>      <span class="co"># Return results from all methods</span></span>
<span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="running-specific-methods">Running Specific Methods<a class="anchor" aria-label="anchor" href="#running-specific-methods"></a>
</h3>
<p>If you want to run only specific methods (e.g., for faster
computation), set <code>run_all = FALSE</code> and specify the
methods:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Run only a subset of methods</span></span>
<span><span class="va">isotwas_model</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_isotwas.html">compute_isotwas</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">snp_mat</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">isoform_mat</span>,</span>
<span>  Y.rep <span class="op">=</span> <span class="va">isoform_mat_rep</span>,</span>
<span>  R <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'mrce_lasso'</span>, <span class="st">'multi_enet'</span>, <span class="st">'sgl'</span>, <span class="st">'univariate'</span><span class="op">)</span>,</span>
<span>  nfolds <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1789</span>,</span>
<span>  run_all <span class="op">=</span> <span class="cn">FALSE</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="using-graph-regularized-regression">Using Graph-Regularized Regression<a class="anchor" aria-label="anchor" href="#using-graph-regularized-regression"></a>
</h3>
<p>The <code>graph_reg</code> method can incorporate prior knowledge
about isoform similarity (e.g., shared exon structure). You can create a
similarity matrix from a GTF annotation file:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create similarity matrix from GTF (based on shared exons)</span></span>
<span><span class="va">sim_result</span> <span class="op">=</span> <span class="fu"><a href="../reference/similarity_from_gtf.html">similarity_from_gtf</a></span><span class="op">(</span></span>
<span>  gtf_path <span class="op">=</span> <span class="st">"gencode.v45.annotation.gtf.gz"</span>,</span>
<span>  gene <span class="op">=</span> <span class="st">"BRCA1"</span>,</span>
<span>  method <span class="op">=</span> <span class="st">"jaccard"</span>  <span class="co"># Options: "jaccard", "overlap_coef", "binary"</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Use in model training</span></span>
<span><span class="va">isotwas_model</span> <span class="op">=</span> <span class="fu"><a href="../reference/compute_isotwas.html">compute_isotwas</a></span><span class="op">(</span></span>
<span>  X <span class="op">=</span> <span class="va">snp_mat</span>,</span>
<span>  Y <span class="op">=</span> <span class="va">isoform_mat</span>,</span>
<span>  Y.rep <span class="op">=</span> <span class="va">isoform_mat_rep</span>,</span>
<span>  R <span class="op">=</span> <span class="fl">10</span>,</span>
<span>  id <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">isoform_mat_rep</span><span class="op">)</span>,</span>
<span>  similarity_matrix <span class="op">=</span> <span class="va">sim_result</span><span class="op">$</span><span class="va">similarity_matrix</span>,</span>
<span>  method <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'multi_enet'</span>, <span class="st">'graph_reg'</span>, <span class="st">'univariate'</span><span class="op">)</span>,</span>
<span>  run_all <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  verbose <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  seed <span class="op">=</span> <span class="fl">1789</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="function-parameters">Function Parameters<a class="anchor" aria-label="anchor" href="#function-parameters"></a>
</h2>
<p>A few notes on the key options:</p>
<ul>
<li><p><strong><code>omega_est</code></strong>: By selecting
<code>omega_est = 'replicates'</code>, we use the bootstrapped
replicates to estimate the correlation between isoforms. The alternative
is <code>omega_est = 'mean'</code>, which will use whatever matrix you
submit for <code>Y</code>.</p></li>
<li><p><strong><code>omega_nlambda</code></strong>: Number of steps
between fully dense to fully sparse in the correlation matrix
estimation. We recommend 5-10 as the default.</p></li>
<li><p><strong><code>method</code></strong>: Vector of methods to use.
By default with <code>run_all = TRUE</code>, all 9 methods are run and
the best model is selected.</p></li>
<li><p><strong><code>similarity_matrix</code></strong>: Optional matrix
for <code>graph_reg</code> method. Can be created using
<code><a href="../reference/similarity_from_gtf.html">similarity_from_gtf()</a></code> or
<code><a href="../reference/create_similarity_from_exons.html">create_similarity_from_exons()</a></code>.</p></li>
<li><p><strong><code>predict_nlambda</code></strong>: Number of penalty
variables for <code>mrce_lasso</code>. We recommend 10-20.</p></li>
<li><p><strong><code>alpha</code></strong>: Elastic net mixing parameter
(0 = ridge, 1 = lasso, 0.5 = elastic net).</p></li>
<li><p><strong><code>verbose</code></strong>: If <code>TRUE</code>
(default), prints detailed progress including per-method timing and
performance.</p></li>
<li><p><strong><code>run_all</code></strong>: If <code>TRUE</code>, all
methods will be run regardless of <code>method</code>
parameter.</p></li>
<li><p><strong><code>return_all</code></strong>: If <code>TRUE</code>,
results from all methods are returned (useful for method
comparison).</p></li>
</ul>
</div>
<div class="section level2">
<h2 id="examining-results">Examining Results<a class="anchor" aria-label="anchor" href="#examining-results"></a>
</h2>
<p>The output from <code><a href="../reference/compute_isotwas.html">compute_isotwas()</a></code> is an
<code>isotwas_result</code> object with several components:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Print summary</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/print.html" class="external-link">print</a></span><span class="op">(</span><span class="va">isotwas_model</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Access the comparison table (R² for each method)</span></span>
<span><span class="va">isotwas_model</span><span class="op">$</span><span class="va">comparison</span></span>
<span></span>
<span><span class="co"># Access the best model</span></span>
<span><span class="va">isotwas_model</span><span class="op">$</span><span class="va">best_models</span></span>
<span></span>
<span><span class="co"># Access individual transcript results</span></span>
<span><span class="va">isotwas_model</span><span class="op">$</span><span class="va">best_models</span><span class="op">$</span><span class="va">Isoform1</span><span class="op">$</span><span class="va">r2</span></span>
<span><span class="va">isotwas_model</span><span class="op">$</span><span class="va">best_models</span><span class="op">$</span><span class="va">Isoform1</span><span class="op">$</span><span class="va">weights</span></span>
<span></span>
<span><span class="co"># Get the full weight matrix</span></span>
<span><span class="va">weight_mat</span> <span class="op">=</span> <span class="fu"><a href="../reference/get_weight_matrix.html">get_weight_matrix</a></span><span class="op">(</span><span class="va">isotwas_model</span><span class="op">$</span><span class="va">best_models</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="converting-the-model">Converting the Model<a class="anchor" aria-label="anchor" href="#converting-the-model"></a>
</h2>
<p>We can convert the model from a list to a tibble or data.frame using
<code><a href="../reference/convert_model.html">convert_model()</a></code>:</p>
<p>Let’s take a look at what a sample model looks like:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_isotwas</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"model_example.RDS"</span>, package <span class="op">=</span> <span class="st">"isotwas"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">class</a></span><span class="op">(</span><span class="va">model_isotwas</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "list"</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">model_isotwas</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] 2</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">model_isotwas</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Model" "R2"</span></span>
<span><span class="va">model_isotwas</span><span class="op">$</span><span class="va">Model</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="co">#&gt; $Transcript</span></span>
<span><span class="co">#&gt; [1] "Isoform1"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Model</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 187 × 2</span></span></span>
<span><span class="co">#&gt;    SNP        Weight</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> SNP10  -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">095</span><span style="color: #BB0000;">7</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> SNP19  -<span style="color: #BB0000;">0.001</span><span style="color: #BB0000; text-decoration: underline;">58</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> SNP80  -<span style="color: #BB0000;">0.003</span><span style="color: #BB0000; text-decoration: underline;">75</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> SNP89  -<span style="color: #BB0000;">0.000</span><span style="color: #BB0000; text-decoration: underline;">884</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> SNP94  -<span style="color: #BB0000;">0.005</span><span style="color: #BB0000; text-decoration: underline;">04</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> SNP124  0.005<span style="text-decoration: underline;">86</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> SNP195 -<span style="color: #BB0000;">0.020</span><span style="color: #BB0000; text-decoration: underline;">1</span>   </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> SNP232 -<span style="color: #BB0000;">0.006</span><span style="color: #BB0000; text-decoration: underline;">27</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> SNP233  0.000<span style="text-decoration: underline;">960</span> </span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> SNP261 -<span style="color: #BB0000;">0.006</span><span style="color: #BB0000; text-decoration: underline;">88</span>  </span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 177 more rows</span></span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $R2</span></span>
<span><span class="co">#&gt; [1] 0.3174829</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $P</span></span>
<span><span class="co">#&gt; value </span></span>
<span><span class="co">#&gt;     0 </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $Pred</span></span>
<span><span class="co">#&gt;       IND190       IND380       IND130       IND314       IND462       IND137 </span></span>
<span><span class="co">#&gt; -0.395051620 -0.405521454 -0.405642239 -0.513143425 -0.219109235 -0.472395293 </span></span>
<span><span class="co">#&gt;       IND191       IND228       IND253       IND358       IND318         IND3 </span></span>
<span><span class="co">#&gt; -0.186011347 -0.047232532 -0.039820719 -0.203023847 -0.279393749 -0.177154706 </span></span>
<span><span class="co">#&gt;        IND28       IND182        IND91       IND476       IND255       IND209 </span></span>
<span><span class="co">#&gt; -0.261468441 -0.572125420 -0.351933854 -0.493579401 -0.417614473 -0.059488250 </span></span>
<span><span class="co">#&gt;       IND135       IND223       IND340       IND120       IND115       IND391 </span></span>
<span><span class="co">#&gt; -0.318086976 -0.319871820 -0.135863999  0.083039071 -0.469286056 -0.281497318 </span></span>
<span><span class="co">#&gt;       IND304       IND163       IND386        IND69       IND121       IND269 </span></span>
<span><span class="co">#&gt; -0.342756842 -0.169790486 -0.552650084 -0.339994140  0.018921426 -0.359762240 </span></span>
<span><span class="co">#&gt;        IND24        IND31       IND349        IND32       IND332       IND495 </span></span>
<span><span class="co">#&gt; -0.171493417 -0.450912106 -0.411439831 -0.220513041 -0.445241556 -0.302148360 </span></span>
<span><span class="co">#&gt;       IND328       IND371       IND272        IND23        IND55       IND463 </span></span>
<span><span class="co">#&gt; -0.572640195 -0.090248595 -0.172602006 -0.107964558 -0.450873553 -0.297293819 </span></span>
<span><span class="co">#&gt;       IND194       IND441       IND113        IND47       IND502       IND409 </span></span>
<span><span class="co">#&gt;  0.002500033 -0.082561510 -0.053926154  0.014315858 -0.263294983 -0.424747059 </span></span>
<span><span class="co">#&gt;        IND82        IND48       IND303       IND176       IND393        IND52 </span></span>
<span><span class="co">#&gt; -0.215259827 -0.180472021 -0.566428240 -0.154369420 -0.292469824 -0.125122866 </span></span>
<span><span class="co">#&gt;       IND458       IND426        IND15        IND10       IND211       IND244 </span></span>
<span><span class="co">#&gt; -0.297137749 -0.571241703 -0.271153164 -0.372562537 -0.139310663  0.058872696 </span></span>
<span><span class="co">#&gt;        IND65        IND50       IND442       IND213       IND157       IND300 </span></span>
<span><span class="co">#&gt; -0.353977257 -0.370280430 -0.369661968 -0.308463656 -0.015128680 -0.372327867 </span></span>
<span><span class="co">#&gt;        IND25       IND254        IND49       IND199       IND189       IND295 </span></span>
<span><span class="co">#&gt; -0.294685131  0.207142198 -0.265891894 -0.210070398 -0.125493031 -0.208084001 </span></span>
<span><span class="co">#&gt;        IND88       IND459       IND339        IND46       IND325       IND419 </span></span>
<span><span class="co">#&gt; -0.188952067 -0.341369214 -0.323699038 -0.194042453 -0.177038660  0.040164417 </span></span>
<span><span class="co">#&gt;       IND302       IND216       IND178       IND282       IND186       IND239 </span></span>
<span><span class="co">#&gt; -0.388742683 -0.144205644 -0.215211141 -0.347436381  0.070054061 -0.016246983 </span></span>
<span><span class="co">#&gt;        IND89        IND71        IND33       IND494       IND369       IND109 </span></span>
<span><span class="co">#&gt; -0.558938422 -0.009956788 -0.518804964 -0.397442503 -0.462158482 -0.620882241 </span></span>
<span><span class="co">#&gt;        IND64       IND364         IND1       IND323        IND60       IND151 </span></span>
<span><span class="co">#&gt; -0.470768744 -0.505031962 -0.474205183 -0.369106906 -0.183303208 -0.474395370 </span></span>
<span><span class="co">#&gt;       IND293       IND437        IND85       IND324       IND375       IND423 </span></span>
<span><span class="co">#&gt; -0.216436666 -0.452684526 -0.437603184 -0.588708106 -0.588703304 -0.443682464 </span></span>
<span><span class="co">#&gt;       IND367       IND516        IND95       IND461       IND451        IND76 </span></span>
<span><span class="co">#&gt; -0.601498157 -0.233798606 -0.410275308 -0.511844894 -0.326213821 -0.370696978 </span></span>
<span><span class="co">#&gt;       IND298       IND310       IND222       IND102       IND491       IND422 </span></span>
<span><span class="co">#&gt; -0.502436692 -0.408828852 -0.111287912 -0.226551260 -0.223498530  0.158679573 </span></span>
<span><span class="co">#&gt;       IND351       IND309       IND277       IND214         IND0       IND438 </span></span>
<span><span class="co">#&gt; -0.344041151 -0.424716100 -0.612379239 -0.231148880 -0.417714624 -0.438855224 </span></span>
<span><span class="co">#&gt;       IND453       IND400       IND443       IND227       IND331       IND169 </span></span>
<span><span class="co">#&gt; -0.594149200 -0.385144435 -0.456708590 -0.532575545 -0.315295504 -0.235696550 </span></span>
<span><span class="co">#&gt;       IND197       IND142       IND416        IND51       IND346        IND27 </span></span>
<span><span class="co">#&gt; -0.298466128  0.034104976 -0.526523901  0.508010660 -0.126042297 -0.128398444 </span></span>
<span><span class="co">#&gt;       IND297       IND479       IND488       IND290        IND44       IND308 </span></span>
<span><span class="co">#&gt; -0.259539348 -0.454747553 -0.504412520 -0.510470021 -0.142340184  0.072124794 </span></span>
<span><span class="co">#&gt;       IND486         IND7       IND284       IND444       IND220       IND374 </span></span>
<span><span class="co">#&gt; -0.329956853 -0.350941227 -0.163496962 -0.413414477 -0.593246582 -0.818095517 </span></span>
<span><span class="co">#&gt;       IND500        IND11       IND354       IND246       IND439       IND392 </span></span>
<span><span class="co">#&gt; -0.501920608 -0.214219476  0.007729149 -0.275404004 -0.644399750 -0.285348795 </span></span>
<span><span class="co">#&gt;       IND448       IND506       IND192       IND484       IND203       IND452 </span></span>
<span><span class="co">#&gt; -0.492513383 -0.310190245 -0.024709405 -0.435482280 -0.269321957 -0.552898172 </span></span>
<span><span class="co">#&gt;        IND92       IND134       IND287       IND424       IND307       IND122 </span></span>
<span><span class="co">#&gt; -0.101055241  0.008905316 -0.306358542 -0.461001763 -0.356879362  0.058371001 </span></span>
<span><span class="co">#&gt;       IND414       IND212        IND20       IND127       IND236       IND258 </span></span>
<span><span class="co">#&gt; -0.311646715 -0.111162671 -0.193471243  0.028880043 -0.407672480  0.142610577 </span></span>
<span><span class="co">#&gt;       IND305       IND268       IND513       IND496       IND147       IND322 </span></span>
<span><span class="co">#&gt; -0.340565157 -0.471679317 -0.404866473 -0.395120760 -0.072070907 -0.277843970 </span></span>
<span><span class="co">#&gt;        IND45        IND68       IND201       IND348        IND43        IND75 </span></span>
<span><span class="co">#&gt;  0.035081966 -0.363338384 -0.652974372 -0.412158936 -0.071163309 -0.080312324 </span></span>
<span><span class="co">#&gt;        IND90        IND41       IND387       IND276       IND465       IND224 </span></span>
<span><span class="co">#&gt;  0.025514564 -0.323200418 -0.477190145 -0.154720858 -0.341964318 -0.196673915 </span></span>
<span><span class="co">#&gt;       IND481       IND143       IND226       IND480       IND126       IND498 </span></span>
<span><span class="co">#&gt; -0.663275058 -0.056703898 -0.082764588 -0.210979906  0.048488465 -0.283945815 </span></span>
<span><span class="co">#&gt;       IND233       IND195        IND61       IND173        IND97       IND478 </span></span>
<span><span class="co">#&gt; -0.266472867 -0.217675991  0.014263476 -0.454493334 -0.550484287 -0.041911623 </span></span>
<span><span class="co">#&gt;        IND18       IND397       IND180       IND315       IND338       IND359 </span></span>
<span><span class="co">#&gt; -0.330665666 -0.436392508 -0.609082779 -0.569036586 -0.212491654 -0.343991921 </span></span>
<span><span class="co">#&gt;       IND508       IND101       IND232       IND454       IND145       IND370 </span></span>
<span><span class="co">#&gt; -0.477263014 -0.464338378 -0.072071562 -0.227516541 -0.389293368 -0.173894914 </span></span>
<span><span class="co">#&gt;       IND159        IND12        IND77       IND421       IND112       IND289 </span></span>
<span><span class="co">#&gt;  0.119232022 -0.094999753 -0.481360824 -0.433878747 -0.100941994 -0.581080319 </span></span>
<span><span class="co">#&gt;       IND353       IND235        IND29       IND221       IND335       IND249 </span></span>
<span><span class="co">#&gt; -0.332866777 -0.078403530 -0.340509458 -0.130391230 -0.223907121 -0.298415252 </span></span>
<span><span class="co">#&gt;        IND84       IND296       IND183       IND344       IND259       IND384 </span></span>
<span><span class="co">#&gt; -0.199496411 -0.410031721 -0.377059571 -0.693998245 -0.370623474 -0.246649423 </span></span>
<span><span class="co">#&gt;       IND148       IND485       IND264        IND54       IND119       IND146 </span></span>
<span><span class="co">#&gt; -0.271335029 -0.225853454 -0.261874967  0.087807798 -0.158416463 -0.214176185 </span></span>
<span><span class="co">#&gt;        IND93        IND80        IND16       IND350       IND503       IND490 </span></span>
<span><span class="co">#&gt; -0.420587615 -0.521984344 -0.070014457 -0.477095206 -0.366749255 -0.412342395 </span></span>
<span><span class="co">#&gt;       IND270        IND70       IND385       IND433       IND428       IND280 </span></span>
<span><span class="co">#&gt; -0.502675568 -0.087020068 -0.312957603 -0.141636588 -0.630821691 -0.190763136 </span></span>
<span><span class="co">#&gt;       IND234       IND104       IND155       IND417       IND207        IND67 </span></span>
<span><span class="co">#&gt; -0.145463209 -0.065338172 -0.102717325 -0.406053957 -0.487929907 -0.542512235 </span></span>
<span><span class="co">#&gt;       IND471         IND6       IND413       IND412       IND245        IND96 </span></span>
<span><span class="co">#&gt;  0.034250790 -0.327396727 -0.497126708 -0.185326538 -0.331678057 -0.078888704 </span></span>
<span><span class="co">#&gt;        IND63       IND301       IND401       IND263       IND352       IND316 </span></span>
<span><span class="co">#&gt; -0.238519359 -0.297359129 -0.424067348 -0.282617895 -0.050625075 -0.255118778 </span></span>
<span><span class="co">#&gt;       IND341       IND489       IND388       IND457       IND181       IND431 </span></span>
<span><span class="co">#&gt; -0.459988827 -0.313308038 -0.311893938  0.040361232 -0.454678575 -0.575772676 </span></span>
<span><span class="co">#&gt;       IND140       IND512       IND330       IND420       IND377       IND139 </span></span>
<span><span class="co">#&gt; -0.218079182 -0.400390897 -0.392403815 -0.258866066 -0.353999148  0.032783885 </span></span>
<span><span class="co">#&gt;       IND171       IND165       IND395       IND436       IND475       IND167 </span></span>
<span><span class="co">#&gt;  0.002110163 -0.083110045 -0.540800075 -0.680934271 -0.444409304 -0.160315170 </span></span>
<span><span class="co">#&gt;       IND111       IND473       IND379       IND501       IND164       IND360 </span></span>
<span><span class="co">#&gt; -0.189364791 -0.415996781 -0.534266922 -0.437482227 -0.269667607 -0.244362503 </span></span>
<span><span class="co">#&gt;       IND390       IND376       IND425        IND66       IND243        IND39 </span></span>
<span><span class="co">#&gt; -0.016687981 -0.300365083 -0.559132052 -0.397677303 -0.253962997 -0.157969029 </span></span>
<span><span class="co">#&gt;       IND125        IND59       IND218       IND361       IND469       IND321 </span></span>
<span><span class="co">#&gt;  0.198755812 -0.270404134 -0.157473118 -0.435929952 -0.198513588 -0.471754809</span></span></code></pre></div>
<p>This model object is a list with a <code>Model</code> and a
<code>R2</code> slot. The <code>Model</code> slot stores a list of
models for each isoform.</p>
<p>We can convert this model from a list to a <code>tibble</code> using
the <code><a href="../reference/convert_model.html">convert_model()</a></code> function. Make sure you have an
annotation object to have position and REF/ALT allele information.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">model_tsv</span> <span class="op">=</span> <span class="fu"><a href="../reference/convert_model.html">convert_model</a></span><span class="op">(</span><span class="va">model_isotwas</span>,</span>
<span>                          snp_annot <span class="op">=</span> <span class="va">snps</span><span class="op">$</span><span class="va">map</span>,</span>
<span>                          snp_var <span class="op">=</span> <span class="st">'marker.ID'</span><span class="op">)</span></span>
<span><span class="va">model_tsv</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># A tibble: 700 × 10</span></span></span>
<span><span class="co">#&gt;    SNP        Weight Transcript    R2  R2.P chromosome genetic.dist physical.pos</span></span>
<span><span class="co">#&gt;    <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>       <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;chr&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span> <span style="color: #949494; font-style: italic;">&lt;dbl&gt;</span>      <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span>        <span style="color: #949494; font-style: italic;">&lt;int&gt;</span></span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 1</span> SNP10    -<span style="color: #BB0000;">9.57</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-5</span> Isoform1   0.317     0          1            0        <span style="text-decoration: underline;">10</span>099</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 2</span> SNP100   -<span style="color: #BB0000;">2.42</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> Isoform3   0.425     0          1            0       <span style="text-decoration: underline;">100</span>107</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 3</span> SNP1009   2.15<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> Isoform3   0.425     0          1            0      1<span style="text-decoration: underline;">009</span>108</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 4</span> SNP101    8.68<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> Isoform2   0.463     0          1            0       <span style="text-decoration: underline;">101</span>086</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 5</span> SNP1012  -<span style="color: #BB0000;">1.99</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> Isoform1   0.317     0          1            0      1<span style="text-decoration: underline;">012</span>093</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 6</span> SNP1017  -<span style="color: #BB0000;">7.31</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> Isoform2   0.463     0          1            0      1<span style="text-decoration: underline;">017</span>093</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 7</span> SNP1020   4.89<span style="color: #949494;">e</span><span style="color: #BB0000;">-3</span> Isoform2   0.463     0          1            0      1<span style="text-decoration: underline;">020</span>103</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 8</span> SNP1025   3.36<span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> Isoform1   0.317     0          1            0      1<span style="text-decoration: underline;">025</span>113</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;"> 9</span> SNP103   -<span style="color: #BB0000;">5.00</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> Isoform2   0.463     0          1            0       <span style="text-decoration: underline;">103</span>078</span></span>
<span><span class="co">#&gt; <span style="color: #BCBCBC;">10</span> SNP1042  -<span style="color: #BB0000;">1.09</span><span style="color: #949494;">e</span><span style="color: #BB0000;">-2</span> Isoform2   0.463     0          1            0      1<span style="text-decoration: underline;">042</span>088</span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 690 more rows</span></span></span>
<span><span class="co">#&gt; <span style="color: #949494;"># ℹ 2 more variables: allele1 &lt;chr&gt;, allele2 &lt;chr&gt;</span></span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="method-recommendations">Method Recommendations<a class="anchor" aria-label="anchor" href="#method-recommendations"></a>
</h2>
<p>Based on simulations and real data analyses:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Default recommendation</strong>: Use
<code>run_all = TRUE</code> to evaluate all methods and let the function
select the best model based on cross-validated R².</p></li>
<li><p><strong>For speed</strong>: If computation time is a concern,
start with <code>c('multi_enet', 'sgl', 'univariate')</code> which
provides a good balance of multivariate and univariate
approaches.</p></li>
<li><p><strong>When isoforms share structure</strong>: Use
<code>graph_reg</code> with a similarity matrix derived from shared
exons. This helps when isoforms have correlated genetic effects due to
shared regulatory regions.</p></li>
<li><p><strong>For robust predictions</strong>: The
<code>stacking</code> method combines multiple base methods and often
provides stable predictions, especially when different methods perform
well for different isoforms.</p></li>
<li><p><strong>For joint feature selection</strong>: Use
<code>mtlasso</code> or <code>sgl</code> when you expect SNPs to affect
multiple isoforms simultaneously.</p></li>
</ol>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Arjun Bhattacharya.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
